<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Second generation methods for immune cell deconvolution • omnideconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Second generation methods for immune cell deconvolution">
<meta property="og:description" content="Collection of second generation methods for immune cell
    deconvolution from bulk RNA-seq samples.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/omnideconv_vignette.html">omnideconv_vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/icbi-lab/omnideconv/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="omnideconv" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#omnideconv" class="anchor"></a>omnideconv</h1></div>
<!-- badges: start -->

<p>The goal of omnideconv is to unify second generation immune deconvolution methods in an R package.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>There are two ways to install it, the long one, installing all packages needed for all methods and the short one, installing only the ones for basic functionality.</p>
<p>If the short install is chosen, the required packages for methods are installed when the methods are used. The same goes for required python packages.</p>
<p>Install the CRAN package devtools and use it to install omnideconv from <a href="https://github.com/">GitHub</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>

<span class="co"># long/complete install</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"icbi-lab/omnideconv"</span>, dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Imports"</span>,
    <span class="st">"Suggests"</span><span class="op">)</span><span class="op">)</span>
<span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/install_all_python.html">install_all_python</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># short/quick install</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"icbi-lab/omnideconv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>The main functions of this package are to build a signature matrix and deconvolute bulk RNA-seq data based on this signature matrix.</p>
<p>The basic workflow is the following:</p>
<div id="1-build-a-signature-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#1-build-a-signature-matrix" class="anchor"></a>1. Build a Signature Matrix</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_data</span>, <span class="va">cell_type_annotations</span>,
    method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></code></pre></div>
<p>In this method, single_cell_data is a matrix of single cell RNA-seq data with genes in rows and cells in columns, cell_type_annotations a vector that contains an annotation for each cell in your single_cell_data, and the possible methods are:</p>
<ul>
<li>AutoGeneS (“autogenes”)</li>
<li>Bisque (“bisque”)</li>
<li>BSeq-sc (“bseqsc”)</li>
<li>CDSeq (“cdseq”)</li>
<li>CibersortX (“cibersortx”)</li>
<li>CPM (“cpm”)</li>
<li>DWLS (“dwls”)</li>
<li>MOMF (“momf”)</li>
<li>MuSiC (“music”)</li>
<li>Scaden (“scaden”)</li>
<li>SCDC (“scdc”)</li>
</ul>
</div>
<div id="2-deconvolute" class="section level3">
<h3 class="hasAnchor">
<a href="#2-deconvolute" class="anchor"></a>2. Deconvolute</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk</span>, <span class="va">signature_matrix</span>, method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></code></pre></div>
<p>Here, bulk is your bulk RNA-seq data as a matrix with genes in rows and samples in column, signature_matrix is the signature matrix you created in the previous step and the method can again be one of the four methods listed above.</p>
<p>This is, what the cell type properties in your bulk RNA-seq data computed in the deconvolution step could look like:</p>
<table class="table">
<colgroup>
<col width="9%">
<col width="6%">
<col width="7%">
<col width="14%">
<col width="16%">
<col width="6%">
<col width="7%">
<col width="7%">
<col width="10%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">B cell</th>
<th align="right">Macrophage</th>
<th align="right">Monocyte conventional</th>
<th align="right">Monocyte non-conventional</th>
<th align="right">NK cell</th>
<th align="right">T cell CD4</th>
<th align="right">T cell CD8</th>
<th align="right">T cell dividing</th>
<th align="right">T cell regulatory</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HD30_PBMC_0</td>
<td align="right">0.0800780</td>
<td align="right">0e+00</td>
<td align="right">0.2764050</td>
<td align="right">0.1427114</td>
<td align="right">0.1385384</td>
<td align="right">0.3316996</td>
<td align="right">0.0277395</td>
<td align="right">0.0026353</td>
<td align="right">0.0001929</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_1</td>
<td align="right">0.0735586</td>
<td align="right">1e-06</td>
<td align="right">0.3554604</td>
<td align="right">0.1236851</td>
<td align="right">0.1253645</td>
<td align="right">0.2968113</td>
<td align="right">0.0243933</td>
<td align="right">0.0007228</td>
<td align="right">0.0000029</td>
</tr>
<tr class="odd">
<td align="left">HD30_PBMC_3</td>
<td align="right">0.0862939</td>
<td align="right">0e+00</td>
<td align="right">0.2591685</td>
<td align="right">0.1246382</td>
<td align="right">0.1305280</td>
<td align="right">0.3595342</td>
<td align="right">0.0360420</td>
<td align="right">0.0037421</td>
<td align="right">0.0000530</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_7</td>
<td align="right">0.1156300</td>
<td align="right">0e+00</td>
<td align="right">0.3316376</td>
<td align="right">0.0226567</td>
<td align="right">0.1140384</td>
<td align="right">0.3596168</td>
<td align="right">0.0545500</td>
<td align="right">0.0017350</td>
<td align="right">0.0001355</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_0</td>
<td align="right">0.1141260</td>
<td align="right">3e-07</td>
<td align="right">0.4713667</td>
<td align="right">0.0024376</td>
<td align="right">0.0511600</td>
<td align="right">0.3026050</td>
<td align="right">0.0571770</td>
<td align="right">0.0010128</td>
<td align="right">0.0001147</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_1</td>
<td align="right">0.1143393</td>
<td align="right">0e+00</td>
<td align="right">0.4559677</td>
<td align="right">0.0000000</td>
<td align="right">0.0408233</td>
<td align="right">0.3235003</td>
<td align="right">0.0651854</td>
<td align="right">0.0001089</td>
<td align="right">0.0000752</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_3</td>
<td align="right">0.0865631</td>
<td align="right">0e+00</td>
<td align="right">0.4423767</td>
<td align="right">0.0000000</td>
<td align="right">0.0486223</td>
<td align="right">0.3520335</td>
<td align="right">0.0686691</td>
<td align="right">0.0015950</td>
<td align="right">0.0001403</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_7</td>
<td align="right">0.0930244</td>
<td align="right">0e+00</td>
<td align="right">0.4262222</td>
<td align="right">0.0000000</td>
<td align="right">0.0583644</td>
<td align="right">0.3587579</td>
<td align="right">0.0632944</td>
<td align="right">0.0000000</td>
<td align="right">0.0003367</td>
</tr>
</tbody>
</table>
</div>
<div id="learn-more" class="section level3">
<h3 class="hasAnchor">
<a href="#learn-more" class="anchor"></a>Learn More</h3>
<p>For more information and an example workflow see the vignette of this package.</p>
</div>
</div>
<div id="requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h2>
<p>Most methods do not require additional software/tokens, but there are a few exceptions:</p>
<ul>
<li>A working version of Docker is required for CibersortX</li>
<li>A token for CibersortX is required from this website: <a href="https://cibersortx.stanford.edu/" class="uri">https://cibersortx.stanford.edu/</a>
</li>
<li>The Cibersort source code is required for BSeq-sc (see tutorial in ?omnideconv::bseqsc_config)</li>
</ul>
</div>
<div id="available-methods-licenses-citations" class="section level2">
<h2 class="hasAnchor">
<a href="#available-methods-licenses-citations" class="anchor"></a>Available methods, Licenses, Citations</h2>
<p>Note that, while <em>omnideconv</em> itself is free (<a href="https://github.com/icbi-lab/omnideconv/blob/main/LICENSE">GPL 3.0</a>), you may need to obtain a license to use the individual methods. See the table below for more information. If you use this package in your work, please cite both our package and the method(s) you are using.</p>
<blockquote>
<p>CITATION</p>
</blockquote>
<table class="table">
<colgroup>
<col width="10%">
<col width="16%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th>method</th>
<th>license</th>
<th>citation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://github.com/theislab/AutoGeneS/">AutoGeneS</a></td>
<td>free (<a href="https://github.com/theislab/AutoGeneS/blob/master/LICENSE">MIT</a>)</td>
<td>Aliee, H., &amp; Theis, F. (2021). AutoGeneS: Automatic gene selection using multi-objective optimization for RNA-seq deconvolution. <a href="https://doi.org/10.1101/2020.02.21.940650" class="uri">https://doi.org/10.1101/2020.02.21.940650</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/cozygene/bisque">Bisque</a></td>
<td>free (<a href="https://github.com/cozygene/bisque/blob/master/DESCRIPTION">GPL 3.0</a>)</td>
<td>Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., &amp; Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. <a href="https://doi.org/10.1038/s41467-020-16607-9" class="uri">https://doi.org/10.1038/s41467-020-16607-9</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/shenorrLab/bseqsc">BSeq-sc</a></td>
<td>free (<a href="https://github.com/shenorrLab/bseqsc/blob/master/DESCRIPTION">GPL 2.0</a>)</td>
<td>Baron, M., Veres, A., Wolock, S. L., Faust, A. L., Gaujoux, R., Vetere, A., Ryu, J. H., Wagner, B. K., Shen-Orr, S. S., Klein, A. M., Melton, D. A., &amp; Yanai, I. (2016). A Single-Cell Transcriptomic Map of the Human and Mouse Pancreas Reveals Inter- and Intra-cell Population Structure. In Cell Systems (Vol. 3, Issue 4, pp. 346–360.e4). <a href="https://doi.org/10.1016/j.cels.2016.08.011" class="uri">https://doi.org/10.1016/j.cels.2016.08.011</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/kkang7/CDSeq_R_Package">CDSeq</a></td>
<td>free (<a href="https://github.com/kkang7/CDSeq_R_Package/blob/master/DESCRIPTION">GPL 3.0</a>)</td>
<td>Kang, K., Huang, C., Li, Y. et al. CDSeqR: fast complete deconvolution for gene expression data from bulk tissues. BMC Bioinformatics 22, 262 (2021). <a href="https://doi.org/10.1186/s12859-021-04186-5" class="uri">https://doi.org/10.1186/s12859-021-04186-5</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://cibersortx.stanford.edu/">CibersortX</a></td>
<td>free for non-commerical use only</td>
<td>Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., Hoang, C. D., Diehn, M., &amp; Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. <a href="https://doi.org/10.1038/nmeth.3337" class="uri">https://doi.org/10.1038/nmeth.3337</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/amitfrish/scBio">CPM</a></td>
<td>free (<a href="https://github.com/amitfrish/scBio/blob/master/DESCRIPTION">GPL 2.0</a>)</td>
<td>Frishberg, A., Peshes-Yaloz, N., Cohn, O., Rosentul, D., Steuerman, Y., Valadarsky, L., Yankovitz, G., Mandelboim, M., Iraqi, F. A., Amit, I., Mayo, L., Bacharach, E., &amp; Gat-Viks, I. (2019). Cell composition analysis of bulk genomics using single-cell data. Nature Methods, 16(4), 327–332. <a href="https://doi.org/10.1038/s41592-019-0355-5" class="uri">https://doi.org/10.1038/s41592-019-0355-5</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://bitbucket.org/yuanlab/dwls/src/master/">DWLS</a></td>
<td>free (<a href="https://bitbucket.org/yuanlab/dwls/src/master/DESCRIPTION">GPL</a>)</td>
<td>Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., &amp; Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. <a href="https://doi.org/10.1038/s41467-019-10802-z" class="uri">https://doi.org/10.1038/s41467-019-10802-z</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/sqsun/MOMF">MOMF</a></td>
<td>free (<a href="https://github.com/sqsun/MOMF/blob/master/LICENSE.md">GPL 3.0</a>)</td>
<td>Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/xuranw/MuSiC/">MuSiC</a></td>
<td>free (<a href="https://github.com/xuranw/MuSiC/blob/master/LICENSE">GPL 3.0</a>)</td>
<td>Wang, X., Park, J., Susztak, K., Zhang, N. R., &amp; Li, M. (2019). Bulk tissue cell type deconvolution with multi-subject single-cell expression reference. Nature Communications, 10(1), 380. <a href="https://doi.org/10.1038/s41467-018-08023-x" class="uri">https://doi.org/10.1038/s41467-018-08023-x</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/KevinMenden/scaden">Scaden</a></td>
<td>free (<a href="https://github.com/KevinMenden/scaden/blob/master/LICENSE">MIT</a>)</td>
<td>Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., &amp; Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. <a href="https://doi.org/10.1101/659227" class="uri">https://doi.org/10.1101/659227</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/meichendong/SCDC">SCDC</a></td>
<td>NA</td>
<td>Dong, M., Thennavan, A., Urrutia, E., Li, Y., Perou, C. M., Zou, F., &amp; Jiang, Y. (2020). SCDC: bulk gene expression deconvolution by multiple single-cell RNA sequencing references. Briefings in Bioinformatics. <a href="https://doi.org/10.1093/bib/bbz166" class="uri">https://doi.org/10.1093/bib/bbz166</a>
</td>
</tr>
</tbody>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/icbi-lab/omnideconv/">https://​github.com/​icbi-lab/​omnideconv/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Konstantin Pelz <br><small class="roles"> Author, maintainer </small>  </li>
<li>Nicolas Goedert <br><small class="roles"> Author </small>  </li>
<li>Yuyu Liang <br><small class="roles"> Author </small>  </li>
<li>Katharina Reinisch <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/icbi-lab/omnideconv/actions"><img src="https://github.com/icbi-lab/omnideconv/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
