<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Second generation methods for immune cell deconvolution • omnideconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Second generation methods for immune cell deconvolution">
<meta property="og:description" content="Collection of second generation methods for immune cell deconvolution from bulk RNA-seq samples">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/omnideconv_vignette.html">omnideconv_vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/icbi-lab/omnideconv/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="omnideconv" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#omnideconv" class="anchor"></a>omnideconv</h1></div>
<!-- badges: start -->

<p>The goal of omnideconv is to unify second generation immune deconvolution methods in an R package.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Install the CRAN package devtools and use it to install omnideconv from <a href="https://github.com/">GitHub</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"icbi-lab/omnideconv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>The main functions of this package are to build a signature matrix and deconvolute bulk RNA-seq data based on this signature matrix.</p>
<p>The basic workflow is the following:</p>
<div id="1-build-a-signature-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#1-build-a-signature-matrix" class="anchor"></a>1. Build a Signature Matrix</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_data</span>, <span class="va">cell_type_annotations</span>, 
    method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></code></pre></div>
<p>In this method, single_cell_data is a matrix of single cell RNA-seq data with genes in rows and cells in columns, cell_type_annotations a vector that contains an annotation for each cell in your single_cell_data, and the possible methods are:</p>
<ul>
<li>Bisque (“bisque”)</li>
<li>DWLS (“dwls”)</li>
<li>MOMF (“momf”)</li>
<li>Scaden (“scaden”)</li>
<li>CibersortX (“cibersortx”)</li>
</ul>
</div>
<div id="2-deconvolute" class="section level3">
<h3 class="hasAnchor">
<a href="#2-deconvolute" class="anchor"></a>2. Deconvolute</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk</span>, <span class="va">signature_matrix</span>, method <span class="op">=</span> <span class="st">"dwls"</span><span class="op">)</span></code></pre></div>
<p>Here, bulk is your bulk RNA-seq data as a matrix with genes in rows and samples in column, signature_matrix is the signature matrix you created in the previous step and the method can again be one of the four methods listed above.</p>
<p>This is, what the cell type properties in your bulk RNA-seq data computed in the deconvolution step could look like:</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="80%">
<col width="1%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">B$ <em>c<strong>e</strong>l<strong>l<em>|</em>M</strong>a<strong>c</strong>r<strong>o</strong>p<strong>h</strong>a<strong>g</strong>e</em>|*M<strong>o</strong>n<strong>o</strong>c<strong>y</strong>t**e* <em>c<strong>o</strong>n<strong>v</strong>e<strong>n</strong>t<strong>i</strong>o<strong>n</strong>a<strong>l<em>|</em>M</strong>o<strong>n</strong>o<strong>c</strong>y<strong>t</strong>e</em> <em>n<strong>o</strong>n</em> − <em>c<strong>o</strong>n<strong>v</strong>e<strong>n</strong>t<strong>i</strong>o<strong>n</strong>a<strong>l<em>|</em>N</strong>K</em> *c<strong>e</strong>l**l<em>|</em>T* *c<strong>e</strong>l**l* *C**D<em>4|</em>T* *c<strong>e</strong>l**l* *C**D<em>8|</em>T* *c<strong>e</strong>l**l* *d<strong>i</strong>v<strong>i</strong>d<strong>i</strong>n**g<em>|</em>T* *c<strong>e</strong>l**l* $regulatory</th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
<th align="right"></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HD30_PBMC_0</td>
<td align="right">0.12816</td>
<td align="right">0</td>
<td align="right">0.51934</td>
<td align="right">0.13378</td>
<td align="right">0.18987</td>
<td align="right">0.00000</td>
<td align="right">0.00561</td>
<td align="right">0</td>
<td align="right">0.02325</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_1</td>
<td align="right">0.09559</td>
<td align="right">0</td>
<td align="right">0.67510</td>
<td align="right">0.00696</td>
<td align="right">0.19640</td>
<td align="right">0.00000</td>
<td align="right">0.00000</td>
<td align="right">0</td>
<td align="right">0.02595</td>
</tr>
<tr class="odd">
<td align="left">HD30_PBMC_3</td>
<td align="right">0.17181</td>
<td align="right">0</td>
<td align="right">0.51368</td>
<td align="right">0.05744</td>
<td align="right">0.16574</td>
<td align="right">0.07258</td>
<td align="right">0.00000</td>
<td align="right">0</td>
<td align="right">0.01876</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_7</td>
<td align="right">0.28474</td>
<td align="right">0</td>
<td align="right">0.44150</td>
<td align="right">0.03053</td>
<td align="right">0.17612</td>
<td align="right">0.05122</td>
<td align="right">0.00000</td>
<td align="right">0</td>
<td align="right">0.01590</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_0</td>
<td align="right">0.08652</td>
<td align="right">0</td>
<td align="right">0.71617</td>
<td align="right">0.00000</td>
<td align="right">0.05495</td>
<td align="right">0.10743</td>
<td align="right">0.02555</td>
<td align="right">0</td>
<td align="right">0.00937</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_1</td>
<td align="right">0.09561</td>
<td align="right">0</td>
<td align="right">0.68077</td>
<td align="right">0.00000</td>
<td align="right">0.04447</td>
<td align="right">0.14894</td>
<td align="right">0.02204</td>
<td align="right">0</td>
<td align="right">0.00818</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_3</td>
<td align="right">0.05961</td>
<td align="right">0</td>
<td align="right">0.71674</td>
<td align="right">0.00000</td>
<td align="right">0.05728</td>
<td align="right">0.12130</td>
<td align="right">0.02990</td>
<td align="right">0</td>
<td align="right">0.01516</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_7</td>
<td align="right">0.06928</td>
<td align="right">0</td>
<td align="right">0.69114</td>
<td align="right">0.00000</td>
<td align="right">0.07708</td>
<td align="right">0.08902</td>
<td align="right">0.04345</td>
<td align="right">0</td>
<td align="right">0.03003</td>
</tr>
</tbody>
</table>
</div>
<div id="learn-more" class="section level3">
<h3 class="hasAnchor">
<a href="#learn-more" class="anchor"></a>Learn More</h3>
<p>For more information and an example workflow see the vignette of this package.</p>
</div>
</div>
<div id="available-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#available-methods" class="anchor"></a>Available Methods</h2>
<p>More information about the second generation deconvolution metods unified in this package as well as their citations can be obtained from the list below.</p>
<ul>
<li>Bisque: Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., &amp; Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. <a href="https://doi.org/10.1038/s41467-020-16607-9" class="uri">https://doi.org/10.1038/s41467-020-16607-9</a>
</li>
<li>CibersortX: Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., Hoang, C. D., Diehn, M., &amp; Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. <a href="https://doi.org/10.1038/nmeth.3337" class="uri">https://doi.org/10.1038/nmeth.3337</a>
</li>
<li>DWLS: Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., &amp; Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. <a href="https://doi.org/10.1038/s41467-019-10802-z" class="uri">https://doi.org/10.1038/s41467-019-10802-z</a>
</li>
<li>MOMF: Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980</li>
<li>Scaden: Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., &amp; Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. <a href="https://doi.org/10.1101/659227" class="uri">https://doi.org/10.1101/659227</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/icbi-lab/omnideconv/">https://​github.com/​icbi-lab/​omnideconv/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/BSD-3-Clause">BSD_3_clause</a> + file <a href="LICENSE-text.html">LICENSE</a>.md</small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Konstantin Pelz <br><small class="roles"> Author, maintainer </small>  </li>
<li>Nicolas Goedert <br><small class="roles"> Author </small>  </li>
<li>Yuyu Liang <br><small class="roles"> Author </small>  </li>
<li>Katharina Reinisch <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/icbi-lab/omnideconv/actions"><img src="https://github.com/icbi-lab/omnideconv/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
