<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Second generation methods for immune cell deconvolution • omnideconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Second generation methods for immune cell deconvolution">
<meta property="og:description" content="Collection of second generation methods for immune cell deconvolution from bulk RNA-seq samples.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/omnideconv_example.html">Deconvolution of the tumor microenvironment with omnideconv</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/omnideconv/omnideconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="omnideconv">omnideconv<a class="anchor" aria-label="anchor" href="#omnideconv"></a>
</h1></div>
<p><a href="https://github.com/omnideconv/omnideconv/actions/workflows/test.yml" class="external-link"><img src="https://github.com/omnideconv/omnideconv/actions/workflows/test.yml/badge.svg" alt="R-CMD-check"></a> <a href="https://github.com/omnideconv/omnideconv/blob/master/LICENSE.md" class="external-link"><img src="https://img.shields.io/badge/license-GPL3-blue.svg" alt="license"></a> <a href="https://omnideconv.github.io/omnideconv"><img src="https://img.shields.io/badge/docs-pkgdown-blue.svg" alt="docs"></a> <a href="https://app.codecov.io/gh/omnideconv/omnideconv?branch=main" class="external-link"><img src="https://codecov.io/gh/omnideconv/omnideconv/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></p>
<p>The goal of omnideconv is to unify second generation cell-type deconvolution methods in an R package.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>There are two ways to install <code>omnideconv</code>:</p>
<ul>
<li>The <em>minimal</em> installation installs only the dependencies required for the basic functionalities. All deconvolution methods need to be installed on-demand.</li>
<li>The <em>complete</em> installation installs all dependencies including all deconvolution methods. This may take a considerable time.</li>
</ul>
<p>Since not all dependencies are on CRAN or Bioconductor, <code>omnideconv</code> is available from GitHub only. We recommend installing it through the <a href="https://github.com/r-lib/pak" class="external-link">pak</a> package manager:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install the `pak` package manager</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"pak"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># minimal installation</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"omnideconv/omnideconv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># complete installation, including Python dependencies</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"omnideconv/omnideconv"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/install_all_python.html">install_all_python</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Upon the first loading, miniconda will be installed if not already present. A dedicated conda environment will be created to host the python-based methods.</p>
</div>
<div class="section level2">
<h2 id="available-methods">Available methods<a class="anchor" aria-label="anchor" href="#available-methods"></a>
</h2>
<p>The methods currently implemented in omnideconv are:</p>
<ul>
<li>AutoGeneS (“autogenes”)</li>
<li>Bisque (“bisque”)</li>
<li>BayesPrism (“bayesprism”)</li>
<li>BSeq-sc (“bseqsc”)</li>
<li>CDSeq (“cdseq”)</li>
<li>CIBERSORTx (“cibersortx”)</li>
<li>CPM (“cpm”)</li>
<li>DWLS (“dwls”)</li>
<li>MOMF (“momf”)</li>
<li>MuSiC (“music”)</li>
<li>Scaden (“scaden”)</li>
<li>SCDC (“scdc”)</li>
</ul>
</div>
<div class="section level2">
<h2 id="general-usage">General usage<a class="anchor" aria-label="anchor" href="#general-usage"></a>
</h2>
<p>All the deconvolution methods included in omnideconv can be run in one step, trough the function <code>deconvolute</code>, which takes in input the matrix of bulk RNAseq to be deconvolved (bulk_gene_expression), along with the training single cell expression matrix (single_cell_object) with the cell type annotations and sample information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconvolution</span> <span class="op">&lt;-</span> <span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk_gene_expression</span>, <span class="va">method</span>,</span>
<span>                                         <span class="va">single_cell_object</span>, <span class="va">cell_type_annotations</span>, <span class="va">batch_ids</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="signature-matrixmodel-building">Signature matrix/model building<a class="anchor" aria-label="anchor" href="#signature-matrixmodel-building"></a>
</h2>
<p>The methods AutoGeneS, BSeq-Sc, DWLS, CIBERSORTx, MOMF and Scaden first optimize their internal model, for example building a signature matrix, and then use this model to perform deconvolution. For these methods, the <code>build_model</code> function can be used. The obtained model can then be given in input to the <code>deconvolute</code> function, omitting the single cell data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_object</span>, <span class="va">cell_type_annotations</span>,</span>
<span>                                     <span class="va">batch_ids</span>, <span class="va">method</span>, <span class="va">bulk_gene_expression</span><span class="op">)</span></span>
<span></span>
<span><span class="va">deconvolution</span> <span class="op">&lt;-</span> <span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk_gene_expression</span>, <span class="va">signature</span><span class="op">)</span></span></code></pre></div>
<p>The <code>deconvolute</code> function returns a sample x cell type matrix with the estimated cell fractions</p>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>Different methods have different requirements in terms of input data. This list has been compiled considering the methods documentation, described data procssing or authors recommendation</p>
<table class="table">
<thead><tr class="header">
<th>Method</th>
<th>Single cell normalization</th>
<th>Bulk normalization</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AutogeneS</td>
<td>CPM</td>
<td>TPM</td>
</tr>
<tr class="even">
<td>BayesPrism</td>
<td>Counts</td>
<td>Counts</td>
</tr>
<tr class="odd">
<td>Bisque</td>
<td>Counts</td>
<td>Counts</td>
</tr>
<tr class="even">
<td>Bseq-Sc</td>
<td>Counts</td>
<td>TPM</td>
</tr>
<tr class="odd">
<td>CDseqR</td>
<td>Counts</td>
<td>Counts</td>
</tr>
<tr class="even">
<td>CIBERSORTx</td>
<td>CPM</td>
<td>TPM</td>
</tr>
<tr class="odd">
<td>CPM</td>
<td>Counts</td>
<td>Counts</td>
</tr>
<tr class="even">
<td>DWLS</td>
<td>Counts</td>
<td>TPM</td>
</tr>
<tr class="odd">
<td>MOMF</td>
<td>Counts</td>
<td>Counts</td>
</tr>
<tr class="even">
<td>MuSiC</td>
<td>Counts</td>
<td>TPM</td>
</tr>
<tr class="odd">
<td>Scaden</td>
<td>Counts</td>
<td>TPM</td>
</tr>
<tr class="even">
<td>SCDC</td>
<td>Counts</td>
<td>TPM</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="learn-more">Learn More<a class="anchor" aria-label="anchor" href="#learn-more"></a>
</h3>
<p>For more information and an example workflow see the vignette of this package.</p>
</div>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<p>Most methods do not require additional software/tokens, but there are a few exceptions:</p>
<ul>
<li>A working version of Docker or Singularity is required for CIBERSORTx</li>
<li>A token for CIBERSORTx is required from this website: <a href="https://cibersortx.stanford.edu/" class="external-link uri">https://cibersortx.stanford.edu/</a>
</li>
<li>The CIBERSORT source code is required for BSeq-sc (see tutorial in ?omnideconv::bseqsc_config)</li>
</ul>
</div>
<div class="section level2">
<h2 id="available-methods-licenses-citations">Available methods, Licenses, Citations<a class="anchor" aria-label="anchor" href="#available-methods-licenses-citations"></a>
</h2>
<p>Note that, while <em>omnideconv</em> itself is free (<a href="https://github.com/omnideconv/omnideconv/blob/main/LICENSE" class="external-link">GPL 3.0</a>), you may need to obtain a license to use the individual methods. See the table below for more information. If you use this package in your work, please cite both our package and the method(s) you are using.</p>
<blockquote>
<p>CITATION</p>
</blockquote>
<table class="table">
<colgroup>
<col width="10%">
<col width="17%">
<col width="72%">
</colgroup>
<thead><tr class="header">
<th>method</th>
<th>license</th>
<th>citation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://github.com/theislab/AutoGeneS/" class="external-link">AutoGeneS</a></td>
<td>free (<a href="https://github.com/theislab/AutoGeneS/blob/master/LICENSE" class="external-link">MIT</a>)</td>
<td>Aliee, H., &amp; Theis, F. (2021). AutoGeneS: Automatic gene selection using multi-objective optimization for RNA-seq deconvolution. <a href="https://doi.org/10.1101/2020.02.21.940650" class="external-link uri">https://doi.org/10.1101/2020.02.21.940650</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/Danko-Lab/BayesPrism" class="external-link">BayesPrism</a></td>
<td>free (<a href="https://github.com/Danko-Lab/BayesPrism/blob/main/BayesPrism/DESCRIPTION" class="external-link">GPL 3.0</a>)</td>
<td>Chu, T., Wang, Z., Pe’er, D. et al. Cell type and gene expression deconvolution with BayesPrism enables Bayesian integrative analysis across bulk and single-cell RNA sequencing in oncology. Nat Cancer 3, 505–517 (2022). <a href="https://doi.org/10.1038/s43018-022-00356-3" class="external-link uri">https://doi.org/10.1038/s43018-022-00356-3</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/cozygene/bisque" class="external-link">Bisque</a></td>
<td>free (<a href="https://github.com/cozygene/bisque/blob/master/DESCRIPTION" class="external-link">GPL 3.0</a>)</td>
<td>Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., &amp; Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. <a href="https://doi.org/10.1038/s41467-020-16607-9" class="external-link uri">https://doi.org/10.1038/s41467-020-16607-9</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/shenorrLab/bseqsc" class="external-link">BSeq-sc</a></td>
<td>free (<a href="https://github.com/shenorrLab/bseqsc/blob/master/DESCRIPTION" class="external-link">GPL 2.0</a>)</td>
<td>Baron, M., Veres, A., Wolock, S. L., Faust, A. L., Gaujoux, R., Vetere, A., Ryu, J. H., Wagner, B. K., Shen-Orr, S. S., Klein, A. M., Melton, D. A., &amp; Yanai, I. (2016). A Single-Cell Transcriptomic Map of the Human and Mouse Pancreas Reveals Inter- and Intra-cell Population Structure. In Cell Systems (Vol. 3, Issue 4, pp. 346–360.e4). <a href="https://doi.org/10.1016/j.cels.2016.08.011" class="external-link uri">https://doi.org/10.1016/j.cels.2016.08.011</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/kkang7/CDSeq_R_Package" class="external-link">CDSeq</a></td>
<td>free (<a href="https://github.com/kkang7/CDSeq_R_Package/blob/master/DESCRIPTION" class="external-link">GPL 3.0</a>)</td>
<td>Kang, K., Huang, C., Li, Y. et al. CDSeqR: fast complete deconvolution for gene expression data from bulk tissues. BMC Bioinformatics 22, 262 (2021). <a href="https://doi.org/10.1186/s12859-021-04186-5" class="external-link uri">https://doi.org/10.1186/s12859-021-04186-5</a>
</td>
</tr>
<tr class="even">
<td><a href="https://cibersortx.stanford.edu/" class="external-link">CIBERSORTx</a></td>
<td>free for non-commerical use only</td>
<td>Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., Hoang, C. D., Diehn, M., &amp; Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. <a href="https://doi.org/10.1038/nmeth.3337" class="external-link uri">https://doi.org/10.1038/nmeth.3337</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/amitfrish/scBio" class="external-link">CPM</a></td>
<td>free (<a href="https://github.com/amitfrish/scBio/blob/master/DESCRIPTION" class="external-link">GPL 2.0</a>)</td>
<td>Frishberg, A., Peshes-Yaloz, N., Cohn, O., Rosentul, D., Steuerman, Y., Valadarsky, L., Yankovitz, G., Mandelboim, M., Iraqi, F. A., Amit, I., Mayo, L., Bacharach, E., &amp; Gat-Viks, I. (2019). Cell composition analysis of bulk genomics using single-cell data. Nature Methods, 16(4), 327–332. <a href="https://doi.org/10.1038/s41592-019-0355-5" class="external-link uri">https://doi.org/10.1038/s41592-019-0355-5</a>
</td>
</tr>
<tr class="even">
<td><a href="https://bitbucket.org/yuanlab/dwls/src/master/" class="external-link">DWLS</a></td>
<td>free (<a href="https://bitbucket.org/yuanlab/dwls/src/master/DESCRIPTION" class="external-link">GPL</a>)</td>
<td>Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., &amp; Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. <a href="https://doi.org/10.1038/s41467-019-10802-z" class="external-link uri">https://doi.org/10.1038/s41467-019-10802-z</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/sqsun/MOMF" class="external-link">MOMF</a></td>
<td>free (<a href="https://github.com/sqsun/MOMF/blob/master/LICENSE.md" class="external-link">GPL 3.0</a>)</td>
<td>Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980</td>
</tr>
<tr class="even">
<td><a href="https://github.com/xuranw/MuSiC/" class="external-link">MuSiC</a></td>
<td>free (<a href="https://github.com/xuranw/MuSiC/blob/master/LICENSE" class="external-link">GPL 3.0</a>)</td>
<td>Wang, X., Park, J., Susztak, K., Zhang, N. R., &amp; Li, M. (2019). Bulk tissue cell type deconvolution with multi-subject single-cell expression reference. Nature Communications, 10(1), 380. <a href="https://doi.org/10.1038/s41467-018-08023-x" class="external-link uri">https://doi.org/10.1038/s41467-018-08023-x</a>
</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/KevinMenden/scaden" class="external-link">Scaden</a></td>
<td>free (<a href="https://github.com/KevinMenden/scaden/blob/master/LICENSE" class="external-link">MIT</a>)</td>
<td>Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., &amp; Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. <a href="https://doi.org/10.1101/659227" class="external-link uri">https://doi.org/10.1101/659227</a>
</td>
</tr>
<tr class="even">
<td><a href="https://github.com/meichendong/SCDC" class="external-link">SCDC</a></td>
<td>(<a href="https://github.com/meichendong/SCDC/blob/master/README.md" class="external-link">MIT</a>)</td>
<td>Dong, M., Thennavan, A., Urrutia, E., Li, Y., Perou, C. M., Zou, F., &amp; Jiang, Y. (2020). SCDC: bulk gene expression deconvolution by multiple single-cell RNA sequencing references. Briefings in Bioinformatics. <a href="https://doi.org/10.1093/bib/bbz166" class="external-link uri">https://doi.org/10.1093/bib/bbz166</a>
</td>
</tr>
</tbody>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/omnideconv/omnideconv/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing omnideconv</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lorenzo Merotto <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0006-0637-8436" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Alexander Dietrich <br><small class="roles"> Author </small>  </li>
<li>Konstantin Pelz <br><small class="roles"> Author </small>  </li>
<li>Nicolas Goedert <br><small class="roles"> Author </small>  </li>
<li>Yuyu Liang <br><small class="roles"> Author </small>  </li>
<li>Katharina Reinisch <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Lorenzo Merotto, Alexander Dietrich, Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
