---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  tidy.opts = list(width.cutoff = 60),
  tidy = TRUE,
  message = FALSE,
  warning = FALSE
)
```

# omnideconv

<!-- badges: start -->
[![R-CMD-check](https://github.com/icbi-lab/omnideconv/workflows/R-CMD-check/badge.svg)](https://github.com/icbi-lab/omnideconv/actions)
<!-- badges: end -->

The goal of omnideconv is to unify second generation immune deconvolution methods in an R package. 

## Installation

There are two ways to install it, the long one installing all packages needed for all methods and
the short one, installing only the ones for basic functionality.
If the short install is chosen, the required packages for methods are installed when the methods
are used.
Install the CRAN package devtools and use it to install omnideconv from [GitHub](https://github.com/):

``` {r installation, eval=FALSE}
install.packages("devtools")

# long/complete install
devtools::install_github("icbi-lab/omnideconv", dependencies = c("Imports", "Suggests"))

# short/quick install
devtools::install_github("icbi-lab/omnideconv")
```
## Usage

The main functions of this package are to build a signature matrix and deconvolute bulk RNA-seq data based on this signature matrix.

The basic workflow is the following:

```{r load-datasets, echo = FALSE}
load("data/bulk.RData")
load("data/cell_type_annotations.RData")
load("data/single_cell_data.RData")
```

### 1. Build a Signature Matrix

```{r signature, eval=FALSE}
omnideconv::build_model(single_cell_data, cell_type_annotations, method = "dwls")
```

In this method, single_cell_data is a matrix of single cell RNA-seq data with genes in rows and cells in columns, cell_type_annotations a vector that contains an annotation for each cell in your single_cell_data, and the possible methods are:

* AutoGeneS ("autogenes")
* Bisque ("bisque")
* BSEQ-sc ("bseqsc")
* CDSeq ("cdseq")
* CibersortX ("cibersortx") 
* CPM ("cpm")
* DWLS ("dwls")
* MOMF ("momf")
* MuSiC ("music")
* Scaden ("scaden")
* SCDC ("scdc")

### 2. Deconvolute 

```{r saveSig, echo=FALSE, results = "hide"}
signature_matrix <- omnideconv::build_model(single_cell_data, cell_type_annotations, "dwls")
```

```{r deconv, eval=FALSE}
omnideconv::deconvolute(bulk, signature_matrix, method = "dwls")
```

Here, bulk is your bulk RNA-seq data as a matrix with genes in rows and samples in column, signature_matrix is the signature matrix you created in the previous step and the method can again be one of the four methods listed above. 

This is, what the cell type properties in your bulk RNA-seq data computed in the deconvolution step could look like:

```{r echo=FALSE, results = "hide"}
res <- omnideconv::deconvolute(bulk, signature_matrix, method = "dwls")
```
```{r echo=FALSE}
knitr::kable(res)
```

### Learn More

For more information and an example workflow see the vignette of this package. 

## Available Methods

More information about the second generation deconvolution methods unified in this package can be obtained from the vignette and the list below. 

* Bisque: Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., & Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. https://doi.org/10.1038/s41467-020-16607-9
* CibersortX: Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., Hoang, C. D., Diehn, M., & Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. https://doi.org/10.1038/nmeth.3337
* DWLS: Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., & Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. https://doi.org/10.1038/s41467-019-10802-z
* MOMF: Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980
* Scaden: Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., & Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. https://doi.org/10.1101/659227
* AutoGeneS: Aliee, H., & Theis, F. (2021). AutoGeneS: Automatic gene selection using multi-objective optimization for RNA-seq deconvolution. https://doi.org/10.1101/2020.02.21.940650
* MuSiC: Wang, X., Park, J., Susztak, K., Zhang, N. R., & Li, M. (2019). Bulk tissue cell type deconvolution with multi-subject single-cell expression reference. Nature Communications, 10(1), 380. https://doi.org/10.1038/s41467-018-08023-x
* SCDC: Dong, M., Thennavan, A., Urrutia, E., Li, Y., Perou, C. M., Zou, F., & Jiang, Y. (2020). SCDC: bulk gene expression deconvolution by multiple single-cell RNA sequencing references. Briefings in Bioinformatics. https://doi.org/10.1093/bib/bbz166
* CPM: Frishberg, A., Peshes-Yaloz, N., Cohn, O., Rosentul, D., Steuerman, Y., Valadarsky, L., Yankovitz, G., Mandelboim, M., Iraqi, F. A., Amit, I., Mayo, L., Bacharach, E., & Gat-Viks, I. (2019). Cell composition analysis of bulk genomics using single-cell data. Nature Methods, 16(4), 327–332. https://doi.org/10.1038/s41592-019-0355-5
* BSEQ-sc ("bseqsc"): Baron, M., Veres, A., Wolock, S. L., Faust, A. L., Gaujoux, R., Vetere, A., Ryu, J. H., Wagner, B. K., Shen-Orr, S. S., Klein, A. M., Melton, D. A., & Yanai, I. (2016). A Single-Cell Transcriptomic Map of the Human and Mouse Pancreas Reveals Inter- and Intra-cell Population Structure. In Cell Systems (Vol. 3, Issue 4, pp. 346–360.e4). https://doi.org/10.1016/j.cels.2016.08.011
* CDSeq: Kang, K., Huang, C., Li, Y. et al. CDSeqR: fast complete deconvolution for gene expression data from bulk tissues. BMC Bioinformatics 22, 262 (2021). https://doi.org/10.1186/s12859-021-04186-5
