---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  tidy.opts=list(width.cutoff=60),
  tidy=TRUE,
  message = FALSE,
  warning = FALSE
)
```

# immunedeconv2

<!-- badges: start -->
[![R-CMD-check](https://github.com/PelzKo/immunedeconv2/workflows/R-CMD-check/badge.svg)](https://github.com/PelzKo/immunedeconv2/actions)
<!-- badges: end -->

The goal of immunedeconv2 is to unify second generation immune deconvolution methods in an R package. 

## Installation

Install the CRAN package devtools and use it to install immunedeconv2 from [GitHub](https://github.com/):

``` {r installation, eval=FALSE}
install.packages("devtools")
devtools::install_github("PelzKo/immunedeconv2")
```
## Usage

The main functions of this package are to build a signature matrix and deconvolute bulk RNA-seq data based on this signature matrix.

The basic workflow is the following:

```{r load datasets, echo = FALSE}
load("data/bulk.RData")
load("data/cell_type_annotations.RData")
load("data/single_cell_data.RData")
```

### 1. Build a Signature Matrix

```{r signature, eval=FALSE}
immunedeconv2::build_model(single_cell_data, cell_type_annotations, method = "dwls")
```

In this method, single_cell_data is a matrix of single cell RNA-seq data with genes in rows and cells in columns, cell_type_annotations a vector that contains an annotation for each cell in your single_cell_data, and the possible methods are:

* Bisque ("bisque")
* DWLS ("dwls")
* MOMF ("momf")
* Scaden ("scaden")

### 2. Deconvolute 

```{r saveSig, echo=FALSE, results = "hide"}
signature_matrix <- immunedeconv2::build_model(single_cell_data, cell_type_annotations, "dwls")
```

```{r deconv, eval=FALSE}
immunedeconv2::deconvolute(bulk, signature_matrix, method = "dwls")
```

Here, bulk is your bulk RNA-seq data as a matrix with genes in rows and samples in column, signature_matrix is the signature matrix you created in the previous step and the method can again be one of the four methods listed above. 

This is, what the cell type properties in your bulk RNA-seq data computed in the deconvolution step could look like:

```{r echo=FALSE, results = "hide"}
res <- immunedeconv2::deconvolute(bulk, signature_matrix, method = "dwls")
```
```{r echo=FALSE}
knitr::kable(res)
```

### Learn More

For more information and an example workflow see the vignette of this package. 

## Available Methods

More information about the second generation deconvolution metods unified in this package as well as their citations can be obtained from the list below. 

* Bisque: Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., PietilÃ¤inen, K. H., Pajukanta, P., & Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. https://doi.org/10.1038/s41467-020-16607-9
* DWLS: Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., & Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. https://doi.org/10.1038/s41467-019-10802-z
* MOMF: Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980
* Scaden: Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., & Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. https://doi.org/10.1101/659227
