<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>SCDC Deconvolution — deconvolute_scdc • omnideconv</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="SCDC Deconvolution — deconvolute_scdc" />
<meta property="og:description" content="This function is to calculate the SCDC deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/omnideconv_vignette.html">omnideconv_vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/icbi-lab/omnideconv/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SCDC Deconvolution</h1>
    <small class="dont-index">Source: <a href='https://github.com/icbi-lab/omnideconv/blob/master/R/SCDC.R'><code>R/SCDC.R</code></a></small>
    <div class="hidden name"><code>deconvolute_scdc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is to calculate the SCDC deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method.</p>
    </div>

    <pre class="usage"><span class='fu'>deconvolute_scdc</span><span class='op'>(</span>
  <span class='va'>bulk_gene_expression</span>,
  <span class='va'>single_cell_object</span>,
  <span class='va'>cell_type_annotations</span>,
  <span class='va'>batch_ids</span>,
  ct_varname <span class='op'>=</span> <span class='st'>"cellType"</span>,
  sample <span class='op'>=</span> <span class='st'>"batchId"</span>,
  ct_sub <span class='op'>=</span> <span class='cn'>NULL</span>,
  iter_max <span class='op'>=</span> <span class='cn'>NULL</span>,
  nu <span class='op'>=</span> <span class='fl'>1e-04</span>,
  epsilon <span class='op'>=</span> <span class='cn'>NULL</span>,
  truep <span class='op'>=</span> <span class='cn'>NULL</span>,
  weight_basis <span class='op'>=</span> <span class='cn'>TRUE</span>,
  ct_cell_size <span class='op'>=</span> <span class='cn'>NULL</span>,
  Transform_bisque <span class='op'>=</span> <span class='cn'>FALSE</span>,
  grid_search <span class='op'>=</span> <span class='cn'>FALSE</span>,
  search_length <span class='op'>=</span> <span class='fl'>0.05</span>,
  names_sc_objects <span class='op'>=</span> <span class='cn'>NULL</span>,
  qcthreshold <span class='op'>=</span> <span class='fl'>0.7</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>,
  quality_control <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bulk_gene_expression</th>
      <td><p>Dataframe or matrix of bulk RNA-seq data (genes x individuals)</p></td>
    </tr>
    <tr>
      <th>single_cell_object</th>
      <td><p>A matrix or dataframe with the single-cell data. Rows are genes,
columns are samples. Row and column names need to be set. This can also be a list of objects,
if SCDC_ENSEMBLE should be used.</p></td>
    </tr>
    <tr>
      <th>cell_type_annotations</th>
      <td><p>A Vector of the cell type annotations. Has to be in the same order
as the samples in single_cell_object. This can also be a list of vectors, if SCDC_ENSEMBLE
should be used.</p></td>
    </tr>
    <tr>
      <th>batch_ids</th>
      <td><p>A vector of the ids of the samples or individuals.</p></td>
    </tr>
    <tr>
      <th>ct_varname</th>
      <td><p>character string specifying the variable name for 'cell types'.</p></td>
    </tr>
    <tr>
      <th>sample</th>
      <td><p>character string specifying the variable name for subject/samples.</p></td>
    </tr>
    <tr>
      <th>ct_sub</th>
      <td><p>vector. a subset of cell types that are selected to construct basis matrix. NULL
means that all are used.</p></td>
    </tr>
    <tr>
      <th>iter_max</th>
      <td><p>the maximum number of iteration in WNNLS. If the parameter is NULL, the default
value of 1000 for a single single cell object and 2000 for a list is chosen.</p></td>
    </tr>
    <tr>
      <th>nu</th>
      <td><p>a small constant to facilitate the calculation of variance</p></td>
    </tr>
    <tr>
      <th>epsilon</th>
      <td><p>a small constant number used for convergence criteria. If the parameter is NULL,
the default value of 0.01 for a single single cell object and 0.001 for a list is chosen.</p></td>
    </tr>
    <tr>
      <th>truep</th>
      <td><p>true cell-type proportions for bulk samples if known</p></td>
    </tr>
    <tr>
      <th>weight_basis</th>
      <td><p>Whether to use the Basis Matrix adjusted for maximal variance weight,
created by the SCDC_basis function</p></td>
    </tr>
    <tr>
      <th>ct_cell_size</th>
      <td><p>default is NULL, which means the "library size" is calculated based on the
data. Users can specify a vector of cell size factors corresponding to the ct.sub according to
prior knowledge. The vector should be named: names(ct_cell_size input) should not be NULL.</p></td>
    </tr>
    <tr>
      <th>Transform_bisque</th>
      <td><p>The bulk sample transformation from bisqueRNA. Aiming to reduce the
systematic difference between single cells and bulk samples.</p></td>
    </tr>
    <tr>
      <th>grid_search</th>
      <td><p>logical. whether to allow grid search method to derive the ENSEMBLE weights.</p></td>
    </tr>
    <tr>
      <th>search_length</th>
      <td><p>a number between 0 to 0.5. if using "Grid search", the step length used.
Smaller search.length derives more accurate optimization results.</p></td>
    </tr>
    <tr>
      <th>names_sc_objects</th>
      <td><p>A vector with the names of the single cell objects. Only used if
a list of single cell objects is supplied. If it remains NULL, the objects are named by their
index.</p></td>
    </tr>
    <tr>
      <th>qcthreshold</th>
      <td><p>The probability threshold used to filter out questionable cells, only used if
quality_control = TRUE.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Whether to create any output.</p></td>
    </tr>
    <tr>
      <th>quality_control</th>
      <td><p>Whether to perform the SCDC_qc quality control method.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Estimated proportion, basis matrix, predicted gene expression levels for bulk samples</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>SCDC_ENSEMBLE can be used by supplying lists to the parameters single_cell_object and
cell_type_annotations. To name the single cell data sets, supply a vector with their
corresponding names to names_sc_objects.</p>
<p>Requires raw read counts.
Works best with multiple cells per single cell patient/subject</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


