<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>CDSeq Deconvolution — deconvolute_cdseq • omnideconv</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="CDSeq Deconvolution — deconvolute_cdseq" />
<meta property="og:description" content="This function is to calculate the CDSeq deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method.
IMPORTANT: The result does not necessarily contain all cell types from the input single cell
data. It assigns cell types to clusters found in the bulk data. See CDSeq_cell_type_assignment_df
for more information." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/omnideconv_vignette.html">Omnideconv Overview</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/icbi-lab/omnideconv/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>CDSeq Deconvolution</h1>
    <small class="dont-index">Source: <a href='https://github.com/icbi-lab/omnideconv/blob/master/R/CDSeq.R'><code>R/CDSeq.R</code></a></small>
    <div class="hidden name"><code>deconvolute_cdseq.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is to calculate the CDSeq deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method.
IMPORTANT: The result does not necessarily contain all cell types from the input single cell
data. It assigns cell types to clusters found in the bulk data. See CDSeq_cell_type_assignment_df
for more information.</p>
    </div>

    <pre class="usage"><span class='fu'>deconvolute_cdseq</span><span class='op'>(</span>
  <span class='va'>bulk_gene_expression</span>,
  <span class='va'>single_cell_object</span>,
  <span class='va'>cell_type_annotations</span>,
  <span class='va'>batch_ids</span>,
  cdseq_gep_sample_specific <span class='op'>=</span> <span class='cn'>NULL</span>,
  batch_correction <span class='op'>=</span> <span class='fl'>1</span>,
  harmony_iter <span class='op'>=</span> <span class='fl'>10</span>,
  harmony_cluster <span class='op'>=</span> <span class='fl'>20</span>,
  nb_size <span class='op'>=</span> <span class='cn'>NULL</span>,
  nb_mu <span class='op'>=</span> <span class='cn'>NULL</span>,
  corr_threshold <span class='op'>=</span> <span class='fl'>0</span>,
  breaksList <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span>, <span class='fl'>0.01</span><span class='op'>)</span>,
  pseudo_cell_count <span class='op'>=</span> <span class='fl'>1</span>,
  seurat_count_threshold <span class='op'>=</span> <span class='fl'>0</span>,
  seurat_scale_factor <span class='op'>=</span> <span class='fl'>10000</span>,
  seurat_norm_method <span class='op'>=</span> <span class='st'>"LogNormalize"</span>,
  seurat_select_method <span class='op'>=</span> <span class='st'>"vst"</span>,
  seurat_nfeatures <span class='op'>=</span> <span class='fl'>1000</span>,
  seurat_npcs <span class='op'>=</span> <span class='fl'>30</span>,
  seurat_dims <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>30</span>,
  seurat_reduction <span class='op'>=</span> <span class='st'>"pca"</span>,
  seurat_resolution <span class='op'>=</span> <span class='fl'>0.8</span>,
  seurat_find_marker <span class='op'>=</span> <span class='cn'>FALSE</span>,
  seurat_DE_test <span class='op'>=</span> <span class='st'>"wilcox"</span>,
  seurat_DE_logfc <span class='op'>=</span> <span class='fl'>0.25</span>,
  seurat_top_n_markers <span class='op'>=</span> <span class='fl'>10</span>,
  sc_pt_size <span class='op'>=</span> <span class='fl'>1</span>,
  cdseq_pt_size <span class='op'>=</span> <span class='fl'>3</span>,
  plot_umap <span class='op'>=</span> <span class='fl'>0</span>,
  plot_tsne <span class='op'>=</span> <span class='fl'>0</span>,
  plot_per_sample <span class='op'>=</span> <span class='fl'>0</span>,
  fig_save <span class='op'>=</span> <span class='fl'>0</span>,
  fig_path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span><span class='op'>(</span><span class='op'>)</span>,
  fig_name <span class='op'>=</span> <span class='st'>"CDSeqCellTypeAssignSCRNA"</span>,
  fig_format <span class='op'>=</span> <span class='st'>"jpeg"</span>,
  fig_dpi <span class='op'>=</span> <span class='fl'>100</span>,
  corr_heatmap_fontsize <span class='op'>=</span> <span class='fl'>10</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bulk_gene_expression</th>
      <td><p>A matrix or dataframe with the bulk data. Rows are genes, columns
are samples.</p></td>
    </tr>
    <tr>
      <th>single_cell_object</th>
      <td><p>A Matrix with the single-cell data. Rows are genes and columns are
samples.</p></td>
    </tr>
    <tr>
      <th>cell_type_annotations</th>
      <td><p>A Vector of the cell type annotations. Has to be in the same order
as the samples in single_cell_object.</p></td>
    </tr>
    <tr>
      <th>batch_ids</th>
      <td><p>A vector of the ids of the samples or individuals.</p></td>
    </tr>
    <tr>
      <th>cdseq_gep_sample_specific</th>
      <td><p>CDSeq-estimated sample-specific cell type gene expression, in
the form of read counts. It is a 3 dimension array, i.e. gene by sample by cell type. The
element cdseq_gep_sample_specific[i,j,k] represents the reads mapped to gene i from cell type
k in sample j.</p></td>
    </tr>
    <tr>
      <th>batch_correction</th>
      <td><p>perform Harmony batch correction if it is 1.</p></td>
    </tr>
    <tr>
      <th>harmony_iter</th>
      <td><p>Maximum number of rounds to run Harmony. One round of Harmony involves one
clustering and one correction step.</p></td>
    </tr>
    <tr>
      <th>harmony_cluster</th>
      <td><p>Maximum number of rounds to run clustering at each round of Harmony.</p></td>
    </tr>
    <tr>
      <th>nb_size</th>
      <td><p>size parameter for negative binomial distribution, check rnbinom for details.</p></td>
    </tr>
    <tr>
      <th>nb_mu</th>
      <td><p>mu parameter for negative binomial distribution, check rnbinom for details.</p></td>
    </tr>
    <tr>
      <th>corr_threshold</th>
      <td><p>if the correlation between CDSeq-estimated GEPs and the scRNAseq GEP is
below this value, then it is considered the two cell types are not matching.</p></td>
    </tr>
    <tr>
      <th>breaksList</th>
      <td><p>parameter for pheatmap controling the color scale. See pheatmap function for
details.</p></td>
    </tr>
    <tr>
      <th>pseudo_cell_count</th>
      <td><p>an integer indicating how many pseudo cells will be generated from
CDSeq-estimated cell-type-specific gene expression profiles. Default values is 1.</p></td>
    </tr>
    <tr>
      <th>seurat_count_threshold</th>
      <td><p>this parameter will be passed to Seurat subset function
(subset = nCount_RNA &gt; seurat_count_threshold) for filtering out single cells whose total
counts is less this threshold.</p></td>
    </tr>
    <tr>
      <th>seurat_scale_factor</th>
      <td><p>this parameter will be passed to scale.factor in Seurat function
NormalizeData.</p></td>
    </tr>
    <tr>
      <th>seurat_norm_method</th>
      <td><p>this parameter will be passed to normalization.method in Seurat
function NormalizeData.</p></td>
    </tr>
    <tr>
      <th>seurat_select_method</th>
      <td><p>this parameter will be passed to selection.method in Seurat
function FindVariableFeatures</p></td>
    </tr>
    <tr>
      <th>seurat_nfeatures</th>
      <td><p>this parameter will be passed to nfeatures in Seurat function
FindVariableFeatures.</p></td>
    </tr>
    <tr>
      <th>seurat_npcs</th>
      <td><p>this parameter will be passed to npcs in Seurat function RunPCA.</p></td>
    </tr>
    <tr>
      <th>seurat_dims</th>
      <td><p>this parameter will be passed to dims in Seurat function FindNeighbors.</p></td>
    </tr>
    <tr>
      <th>seurat_reduction</th>
      <td><p>this parameter will be passed to reduction in Seurat function
FindNeighbors.</p></td>
    </tr>
    <tr>
      <th>seurat_resolution</th>
      <td><p>this parameter will be passed to resolution in Seurat function
FindClusters.</p></td>
    </tr>
    <tr>
      <th>seurat_find_marker</th>
      <td><p>this parameter controls if run seurat FindMarker function, default
is FALSE.</p></td>
    </tr>
    <tr>
      <th>seurat_DE_test</th>
      <td><p>this parameter will be passed to test.use in Seurat function
FindAllMarkers.</p></td>
    </tr>
    <tr>
      <th>seurat_DE_logfc</th>
      <td><p>this parameter will be passed to logfc.threshold in Seurat function
FindAllMarkers.</p></td>
    </tr>
    <tr>
      <th>seurat_top_n_markers</th>
      <td><p>the number of top DE markers saved from Seurat output.</p></td>
    </tr>
    <tr>
      <th>sc_pt_size</th>
      <td><p>point size of single cell data in umap and tsne plots</p></td>
    </tr>
    <tr>
      <th>cdseq_pt_size</th>
      <td><p>point size of CDSeq-estimated cell types in umap and tsne plots</p></td>
    </tr>
    <tr>
      <th>plot_umap</th>
      <td><p>set 1 to plot umap figure of scRNAseq and CDSeq-estimated cell types,
0 otherwise.</p></td>
    </tr>
    <tr>
      <th>plot_tsne</th>
      <td><p>set 1 to plot tsne figure of scRNAseq and CDSeq-estimated cell types,
0 otherwise.</p></td>
    </tr>
    <tr>
      <th>plot_per_sample</th>
      <td><p>currently disabled for debugging</p></td>
    </tr>
    <tr>
      <th>fig_save</th>
      <td><p>1 or 0. 1 means save figures to local and 0 means do not save figures to local.</p></td>
    </tr>
    <tr>
      <th>fig_path</th>
      <td><p>the location where the heatmap figure is saved.</p></td>
    </tr>
    <tr>
      <th>fig_name</th>
      <td><p>the name of umap and tsne figures. Umap figure will have the name of
fig_name_umap_date and tsne figure will be named fig_name_tsne_date.</p></td>
    </tr>
    <tr>
      <th>fig_format</th>
      <td><p>"pdf", "jpeg", or "png".</p></td>
    </tr>
    <tr>
      <th>fig_dpi</th>
      <td><p>figure dpi</p></td>
    </tr>
    <tr>
      <th>corr_heatmap_fontsize</th>
      <td><p>font size of the correlation heatmap between scRNAseq GEP and
CDSeq-estimated GEPs.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Whether to produce an output on the console</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list including:</p>
<dt>fig_path</dt><dd><p>The same as the input fig_path.</p></dd>
<dt>fig_name</dt><dd><p>The same as the input fig_name.</p></dd>
<dt>cdseq_synth_scRNA</dt><dd><p>The synthetic scRNAseq data generated using CDSeq-estiamted GEPs.</p></dd>
<dt>cdseq_scRNA_umap</dt><dd><p>A ggplot figure of the umap outcome.</p></dd>
<dt>cdseq_scRNA_tsne</dt><dd><p>A ggplot figure of the tsne outcome.</p></dd>
<dt>cdseq_synth_scRNA_seurat</dt><dd><p>A Seurat object containing the scRNAseq combined with
CDSeq-estimated cell types. Cell id for CDSeq-estimated cell types start with "CDSeq".</p></dd>
<dt>seurat_cluster_purity</dt><dd><p>For all cells in a Seurat cluster i,  the ith value in
seurat_cluster_purity is the proportion of the mostly repeated cell annotation from
sc_annotation.
For example, after Seurat clustering, suppose there are 100 cells in cluster 1, out of these
100 cells, 90 cells' annotation in sc_annotation is cell type A, then the fist value in
seurat_cluster_purity is 0.9. This output can be used to assess the agreement between
Seurat clustering and the given sc_annotation.</p></dd>
<dt>seurat_unique_clusters</dt><dd><p>Unique Seurat cluster numbering. This can be used together with
seurat_cluster_gold_label to match the Seurat clusters with given annotations.</p></dd>
<dt>seurat_cluster_gold_label</dt><dd><p>The cell type annotations for each unique Seurat cluster based
on sc_annotation.</p></dd>
<dt>seurat_markers</dt><dd><p>DE genes for each Seurat cluster.</p></dd>
<dt>seurat_top_markers</dt><dd><p>Top seurat_top_n_markers DE genes for each Seurat cluster.</p></dd>
<dt>CDSeq_cell_type_assignment_df</dt><dd><p>The cell type assignment for CDSeq-estimated cell types.</p></dd>
<dt>CDSeq_cell_type_assignment_confidence</dt><dd><p>The cell type assignment confidence matrix, only
available when pseudo_cell_count &gt; 1.</p></dd>
<dt>CDSeq_cell_type_assignment_df_all</dt><dd><p>The cell type assignment for CDSeq-estimated cell types,
only available when pseudo_cell_count &gt; 1.</p></dd>
<dt>cdseq_prop_merged</dt><dd><p>CDSeq-estimated cell type proportions with cell type annotations
(annotated using clustering with scRNAseq).</p></dd>
<dt>cdseq_gep_sample_specific_merged</dt><dd><p>Sample-specific cell-type read counts. It is a 3d array
with dimensions: gene, sample, cell type.</p></dd>
<dt>input_list</dt><dd><p>The values of the input parameters.</p></dd>
<dt>cdseq_sc_comb_umap_df</dt><dd><p>The dataframe for umap plot.</p></dd>
<dt>cdseq_sc_comb_tsne_df</dt><dd><p>The dataframe for tsne plot.</p></dd>
<dt>cdseq_prop_merged_byCorr</dt><dd><p>CDSeq-estimated cell type proportions with cell type annotations
(annotated using correlation with scRNAseq).</p></dd>
<dt>cdseq_gep_merged_byCorr</dt><dd><p>CDSeq-estimated cell-type-specific GEPs with cell type annotations
(annotated using correlation with scRNAseq).</p></dd>
<dt>cdseq_annotation_byCorr</dt><dd><p>CDSeq-estimated cell type annotations (annotated using correlation
with scRNAseq).</p></dd>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


