<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Bayesian deconvolution module using BayesPrism — deconvolute_bayesprism • omnideconv</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Bayesian deconvolution module using BayesPrism — deconvolute_bayesprism"><meta property="og:description" content="IMPORTANT: no model is needed. Everything is done inside this method."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/omnideconv_example.html">Deconvolution of the tumor microenvironment with omnideconv</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/omnideconv/omnideconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bayesian deconvolution module using BayesPrism</h1>
    <small class="dont-index">Source: <a href="https://github.com/omnideconv/omnideconv/blob/HEAD/R/BayesPrism.R" class="external-link"><code>R/BayesPrism.R</code></a></small>
    <div class="hidden name"><code>deconvolute_bayesprism.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>IMPORTANT: no model is needed. Everything is done inside this method.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">deconvolute_bayesprism</span><span class="op">(</span></span>
<span>  <span class="va">bulk_gene_expression</span>,</span>
<span>  <span class="va">single_cell_object</span>,</span>
<span>  <span class="va">cell_type_annotations</span>,</span>
<span>  cell_subtype_labels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tum_key <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  apply_bayes_prism_filtering <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"hs"</span>,</span>
<span>  exp.cells <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  pseudo_min <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  gene_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"other_Rb"</span>, <span class="st">"chrM"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>, <span class="st">"Rb"</span>, <span class="st">"Mrp"</span>, <span class="st">"act"</span>, <span class="st">"hb"</span>, <span class="st">"MALAT1"</span><span class="op">)</span>,</span>
<span>  outlier_cut <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  outlier_fraction <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  update_gibbs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gibbs_control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>chain.length <span class="op">=</span> <span class="fl">1000</span>, burn.in <span class="op">=</span> <span class="fl">500</span>, thinning <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  opt_control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>trace <span class="op">=</span> <span class="fl">0</span>, maxit <span class="op">=</span> <span class="fl">1e+05</span><span class="op">)</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  which_theta <span class="op">=</span> <span class="st">"final"</span>,</span>
<span>  state_or_type <span class="op">=</span> <span class="st">"type"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bulk_gene_expression</dt>
<dd><p>A matrix of bulk data. Rows are genes, columns are samples.
Row and column names need to be set.</p></dd>


<dt>single_cell_object</dt>
<dd><p>A matrix with the single-cell data. Rows are genes, columns are
samples. Row and column names need to be set.</p></dd>


<dt>cell_type_annotations</dt>
<dd><p>A vector of the cell type annotations. Has to be in the same order
as the samples in single_cell_object.</p></dd>


<dt>cell_subtype_labels</dt>
<dd><p>a character or factor vector indicating the cell subtype of each row of
the raw count matrix of scRNA-seq or gene expression profile (GEP). The length needs be equal to
nrow(ref.dat). Default is NULL, which uses the same value of cell.type.labels. Note that TED
computes the posterior sum over the subtypes to get the total fraction / expression of each cell type.
This allows a more fine-grained definition of cell types / cell states.</p></dd>


<dt>tum_key</dt>
<dd><p>The character in cell.type.labels denoting the tumor cells, e.g. "tumor" or "malignant".</p></dd>


<dt>apply_bayes_prism_filtering</dt>
<dd><p>set to TRUE if you want to run the <code>cleanup.genes</code> function of BayesPrism; default is FALSE</p></dd>


<dt>species</dt>
<dd><p>A character variable to denote if genes are human ("mm") or mouse ("hs").</p></dd>


<dt>exp.cells</dt>
<dd><p>Genes expressed in number of cells fewer than this will be excluded. Default=1. If the input is GEP, gene
will be selected by automatically setting exp.cells is set to min(exp.cells,1). As a result genes expressed in at least 0
or 1 cell type will be retained. Only used when <code>apply_bayes_prism_filtering</code> is TRUE.</p></dd>


<dt>pseudo_min</dt>
<dd><p>A numeric value indicating the minimum (non-zero) value of phi. Default=1E-8.</p></dd>


<dt>gene_group</dt>
<dd><p>a character vector to input gene groups to be removed, must be one or more elements from
c("other_Rb","chrM","chrX","chrY","Rb","Mrp","act","hb","MALAT1"). Only used when <code>apply_bayes_prism_filtering</code> is TRUE.</p></dd>


<dt>outlier_cut, outlier_fraction</dt>
<dd><p>Filter genes in X whose expression fraction is greater than outlier.cut
(Default=0.01) in more than outlier.fraction (Default=0.1) of bulk data. Typically for dataset with reasonable
quality control, very few genes will be filtered. Removal of outlier genes will ensure that the inference
will not be dominated by outliers, which sometimes may be resulted from poor QC in mapping.</p></dd>


<dt>update_gibbs</dt>
<dd><p>A logical variable to denote whether run final Gibbs sampling to update theta. Default=TRUE.</p></dd>


<dt>gibbs_control</dt>
<dd><p>A list of parameters controlling the Gibbs sampling. Default chain.length=1000,
burn.in=500, thinning=2. A list of parameters controlling the Gibbs sampling. Default chain.length=1000,
burn.in=500, thinning=2. Previous version default is chain.length=400, burn.in=200, thinning=2.
Default chain length has been increased to accommodate spatial transcriptomic data which usually
has lower depth than conventional bulk data, and hence may need longer chain to reach the stationary distribution.</p></dd>


<dt>opt_control</dt>
<dd><p>A list of parameters controlling the optimization by Rcgmin, Default trace=0, maxit= 100000.</p></dd>


<dt>n_cores</dt>
<dd><p>Number of CPU threads used for parallel computing. Default=1</p></dd>


<dt>which_theta</dt>
<dd><p>A character variable to denote whether to extract results from first or final Gibbs sampling.</p></dd>


<dt>state_or_type</dt>
<dd><p>A character variable to extract results from cell type or cell state. We caution the
interpretation of cell states information when their transcription are highly co-linear.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of results is returned including:</p><ul><li><p>bp.resThe result of <code>run.prism</code>, a "BayesPrism" S4 object.</p></li>
<li><p>thetaThe result of <code>get.fraction</code>, the extracted cell fraction results from the BayesPrism object.</p></li>
<li><p>bp.res$prismAn S4 object of the class "prism" to represent the input prism object;</p></li>
<li><p>bp.res$posterior.initial.cellStateAn S4 object of the class "jointPost" to represent the posterior mean of cell state fraction and cell state expression outputted by the initial Gibbs sampling using cell state pirors. Contains Z (inferred expression), theta (inferred fraction) and theta.cv (coefficient of variation of posterior of theta)</p></li>
<li><p>bp.res$posterior.initial.cellTypeAn S4 object of the class "jointPost" to represent the posterior sum of cell states from each cell type (posterior.initial.cellState);</p></li>
<li><p>bp.res$reference.updateAn S4 obejct of the class "reference" to represent the updated profile ψ;</p></li>
<li><p>bp.res$posterior.theta_f:An S4 object of the class "thetaPost" to represent the updated cell type fraction. Contains theta (inferred fraction) and theta.cv (coefficient of variation of posterior of theta);</p></li>
<li><p>bpres$control_paramA list storing the gibbs.control, opt.control and update.gibbs arguments.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Run Bayesian deconvolution to estimate cell type composition and gene expression.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lorenzo Merotto, Alexander Dietrich, Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

