<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SCDC Deconvolution — deconvolute_scdc • omnideconv</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="SCDC Deconvolution — deconvolute_scdc"><meta property="og:description" content="This function is to calculate the SCDC deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/omnideconv_vignette.html">Omnideconv Overview</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/omnideconv/omnideconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SCDC Deconvolution</h1>
    <small class="dont-index">Source: <a href="https://github.com/omnideconv/omnideconv/blob/HEAD/R/SCDC.R" class="external-link"><code>R/SCDC.R</code></a></small>
    <div class="hidden name"><code>deconvolute_scdc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is to calculate the SCDC deconvolution proportions.
IMPORTANT: No model is needed. Everything is done inside this method.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">deconvolute_scdc</span><span class="op">(</span></span>
<span>  <span class="va">bulk_gene_expression</span>,</span>
<span>  <span class="va">single_cell_object</span>,</span>
<span>  <span class="va">cell_type_annotations</span>,</span>
<span>  <span class="va">batch_ids</span>,</span>
<span>  ct_varname <span class="op">=</span> <span class="st">"cellType"</span>,</span>
<span>  sample <span class="op">=</span> <span class="st">"batchId"</span>,</span>
<span>  ct_sub <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  iter_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nu <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  epsilon <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  truep <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  weight_basis <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ct_cell_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform_bisque <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  grid_search <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  search_length <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  names_sc_objects <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  qcthreshold <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quality_control <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bulk_gene_expression</dt>
<dd><p>A matrix of bulk data. Rows are genes, columns are samples.
Row and column names need to be set.</p></dd>


<dt>single_cell_object</dt>
<dd><p>A matrix or dataframe with the single-cell data. Rows are genes,
columns are samples. Row and column names need to be set. This can also be a list of objects,
if SCDC_ENSEMBLE should be used.</p></dd>


<dt>cell_type_annotations</dt>
<dd><p>A Vector of the cell type annotations. Has to be in the same order
as the samples in single_cell_object. This can also be a list of vectors, if SCDC_ENSEMBLE
should be used.</p></dd>


<dt>batch_ids</dt>
<dd><p>A vector of the ids of the samples or individuals.</p></dd>


<dt>ct_varname</dt>
<dd><p>character string specifying the variable name for 'cell types'.</p></dd>


<dt>sample</dt>
<dd><p>character string specifying the variable name for subject/samples.</p></dd>


<dt>ct_sub</dt>
<dd><p>vector. a subset of cell types that are selected to construct basis matrix. NULL
means that all are used.</p></dd>


<dt>iter_max</dt>
<dd><p>the maximum number of iteration in WNNLS. If the parameter is NULL, the default
value of 1000 for a single single cell object and 2000 for a list is chosen.</p></dd>


<dt>nu</dt>
<dd><p>a small constant to facilitate the calculation of variance.</p></dd>


<dt>epsilon</dt>
<dd><p>a small constant number used for convergence criteria. If the parameter is NULL,
the default value of 0.01 for a single single cell object and 0.001 for a list is chosen.</p></dd>


<dt>truep</dt>
<dd><p>true cell-type proportions for bulk samples if known.</p></dd>


<dt>weight_basis</dt>
<dd><p>Whether to use the Basis Matrix adjusted for maximal variance weight,
created by the SCDC_basis function.</p></dd>


<dt>ct_cell_size</dt>
<dd><p>default is NULL, which means the "library size" is calculated based on the
data. Users can specify a vector of cell size factors corresponding to the ct.sub according to
prior knowledge. The vector should be named: names(ct_cell_size input) should not be NULL.</p></dd>


<dt>transform_bisque</dt>
<dd><p>The bulk sample transformation from bisqueRNA. Aiming to reduce the
systematic difference between single cells and bulk samples.</p></dd>


<dt>grid_search</dt>
<dd><p>logical. whether to allow grid search method to derive the ENSEMBLE weights.</p></dd>


<dt>search_length</dt>
<dd><p>a number between 0 to 0.5. if using "Grid search", the step length used.
Smaller search.length derives more accurate optimization results.</p></dd>


<dt>names_sc_objects</dt>
<dd><p>A vector with the names of the single cell objects. Only used if
a list of single cell objects is supplied. If it remains NULL, the objects are named by their
index.</p></dd>


<dt>qcthreshold</dt>
<dd><p>The probability threshold used to filter out questionable cells, only used if
quality_control = TRUE.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to produce an output on the console.</p></dd>


<dt>quality_control</dt>
<dd><p>Whether to perform the SCDC_qc quality control method.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Depends on whether one or multiple single cell sets are used.<br><br></p>


<p>One:</p>
<dl><dt>prop.est.mvw</dt>
<dd><p>A matrix of cell type proportion estimates with cell types as rows and
individuals as columns.</p></dd>

<dt>basis.mvw</dt>
<dd><p>The signature matrix. Rows are genes, columns are cell types.</p></dd>

<dt>yhat</dt>
<dd><p>The predicted gene expression levels for the bulk samples.</p></dd>

<dt>yeval</dt>
<dd><p>The evaluation of the predicted gene expression levels.</p></dd>

<dt>peval</dt>
<dd><p>The evaluation of the deconvoluted proportions. Since we dont have a ground truth,
this is always NULL.</p></dd>


</dl><p>ENSEMBLE:</p>
<dl><dt>w_table</dt>
<dd><p>A matrix with the suggested weights for each single cell dataset and some
statistical evaluation.</p></dd>

<dt>prop.list</dt>
<dd><p>A list of the "One:" outputs as seen above for each single cell dataset.</p></dd>

<dt>prop.only</dt>
<dd><p>A list of the prop.est.mvw values for each single cell dataset.</p></dd>

<dt>gridres</dt>
<dd><p>A matrix with the results of the gridsearch. NULL if grid_search = FALSE.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>SCDC_ENSEMBLE can be used by supplying lists to the parameters single_cell_object and
cell_type_annotations. To name the single cell data sets, supply a vector with their
corresponding names to names_sc_objects.</p>
<p>Requires raw read counts.
Works best with multiple cells per single cell patient/subject</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

