<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ omnideconv</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/omnideconv_vignette.html">Omnideconv Overview</a>
      <a href="../articles/omnideconv_tumor_example.html">Omnideconv Example</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/omnideconv/omnideconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="main-package-methods">Main package methods <a href="#main-package-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Main methods to build signature matrices and deconvolute cell fractions.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolution_methods.html">deconvolution_methods</a></code> </p>
        </td>
        <td><p>List of supported deconvolution methods</p></td>
      </tr><tr><td>
          <p><code><a href="build_model.html">build_model()</a></code> </p>
        </td>
        <td><p>Building the signature matrix</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute.html">deconvolute()</a></code> </p>
        </td>
        <td><p>Deconvolution</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="signature-matrix-creation-methods">Signature matrix creation methods <a href="#signature-matrix-creation-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Direct access to the individual signature matrix creation methods.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="build_model_autogenes.html">build_model_autogenes()</a></code> </p>
        </td>
        <td><p>Calculates the signature model with AutoGeneS</p></td>
      </tr><tr><td>
          <p><code><a href="extract_signature_autogenes.html">extract_signature_autogenes()</a></code> </p>
        </td>
        <td><p>Extraction of the signature matrix from autogenes .pickle files</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_bayesprism.html">build_model_bayesprism()</a></code> </p>
        </td>
        <td><p>No model is built as BayesPrism performs signature building and deconvolution in one step.</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_bisque.html">build_model_bisque()</a></code> </p>
        </td>
        <td><p>No model is build as Bisque does both steps in one.</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_bseqsc.html">build_model_bseqsc()</a></code> </p>
        </td>
        <td><p>Calculates the signature model with BSeq-sc</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_cibersortx.html">build_model_cibersortx()</a></code> </p>
        </td>
        <td><p>Signature matrix creation with CIBERSORTx</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_cpm.html">build_model_cpm()</a></code> </p>
        </td>
        <td><p>No model is build as CPM does not use a signature matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_dwls.html">build_model_dwls()</a></code> </p>
        </td>
        <td><p>Signature matrix creation with DWLS using genes identified by a differential analysis</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_momf.html">build_model_momf()</a></code> </p>
        </td>
        <td><p>Calculates the signature model with MOMF</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_music.html">build_model_music()</a></code> </p>
        </td>
        <td><p>Signature creation with MuSiC.</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_scaden.html">build_model_scaden()</a></code> </p>
        </td>
        <td><p>Builds Scaden Model from scRNA data</p></td>
      </tr><tr><td>
          <p><code><a href="build_model_scdc.html">build_model_scdc()</a></code> </p>
        </td>
        <td><p>Signature creation with SCDC.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="deconvolution-methods">Deconvolution methods <a href="#deconvolution-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Direct access to the individual deconvolution methods to access special features that are not available through <code>deconvolute</code>.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="deconvolute_autogenes.html">deconvolute_autogenes()</a></code> </p>
        </td>
        <td><p>Deconvolution Analysis using AutoGeneS</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_bayesprism.html">deconvolute_bayesprism()</a></code> </p>
        </td>
        <td><p>Bayesian deconvolution module using BayesPrism</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_bisque.html">deconvolute_bisque()</a></code> </p>
        </td>
        <td><p>Calculates the decomposition using the bisque algorithm
IMPORTANT: No model is needed. Everything is done inside this method.</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_bseqsc.html">deconvolute_bseqsc()</a></code> </p>
        </td>
        <td><p>Deconvolution Analysis using BSeq-sc</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_cibersortx.html">deconvolute_cibersortx()</a></code> </p>
        </td>
        <td><p>Deconvolute with CIBERSORTx</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_cpm.html">deconvolute_cpm()</a></code> </p>
        </td>
        <td><p>CPM Deconvolution</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_dwls.html">deconvolute_dwls()</a></code> </p>
        </td>
        <td><p>Calculates the decomposition using the dwls algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_momf.html">deconvolute_momf()</a></code> </p>
        </td>
        <td><p>Deconvolution Analysis using MOMF (via Nonnegative Factorization)</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_music.html">deconvolute_music()</a></code> </p>
        </td>
        <td><p>MuSiC Deconvolution</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_scaden.html">deconvolute_scaden()</a></code> </p>
        </td>
        <td><p>Performs deconvolution with Scaden</p></td>
      </tr><tr><td>
          <p><code><a href="deconvolute_scdc.html">deconvolute_scdc()</a></code> </p>
        </td>
        <td><p>SCDC Deconvolution</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="environment">Environment <a href="#environment" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Methods needed to set credentials or change variables in the environment.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="set_cibersortx_credentials.html">set_cibersortx_credentials()</a></code> </p>
        </td>
        <td><p>Setting the CIBERSORTx Credentials</p></td>
      </tr><tr><td>
          <p><code><a href="bseqsc_config.html">bseqsc_config()</a></code> </p>
        </td>
        <td><p>Setup BSeq-SC External Dependencies</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="datasets">Datasets <a href="#datasets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>List of datasets provided by the package.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bulk.html">bulk</a></code> </p>
        </td>
        <td><p>Example RNA-seq dataset from Hoek et al.</p></td>
      </tr><tr><td>
          <p><code><a href="RefData.html">RefData</a></code> </p>
        </td>
        <td><p>Gold standard measurements with FACS from Hoek et al.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_type_annotations_1.html">cell_type_annotations_1</a></code> </p>
        </td>
        <td><p>Cell Type Annotations Number 1 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_type_annotations_2.html">cell_type_annotations_2</a></code> </p>
        </td>
        <td><p>Cell Type Annotations Number 2 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_type_annotations_3.html">cell_type_annotations_3</a></code> </p>
        </td>
        <td><p>Cell Type Annotations Number 3 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="single_cell_data_1.html">single_cell_data_1</a></code> </p>
        </td>
        <td><p>Example single-cell RNA-seq dataset Number 1 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="single_cell_data_2.html">single_cell_data_2</a></code> </p>
        </td>
        <td><p>Example single-cell RNA-seq dataset Number 2 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="single_cell_data_3.html">single_cell_data_3</a></code> </p>
        </td>
        <td><p>Example single-cell RNA-seq dataset Number 3 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="batch_ids_1.html">batch_ids_1</a></code> </p>
        </td>
        <td><p>Patient ids Number 1 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="batch_ids_2.html">batch_ids_2</a></code> </p>
        </td>
        <td><p>Patient ids Number 2 from Hao et al.</p></td>
      </tr><tr><td>
          <p><code><a href="batch_ids_3.html">batch_ids_3</a></code> </p>
        </td>
        <td><p>Patient ids Number 3 from Hao et al.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualisation">Visualisation <a href="#visualisation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Methods which produce plots with different methods.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="make_benchmarking_scatterplot.html">make_benchmarking_scatterplot()</a></code> </p>
        </td>
        <td><p>Make a Scatterplot for Benchmarking</p></td>
      </tr><tr><td>
          <p><code><a href="make_barplot.html">make_barplot()</a></code> </p>
        </td>
        <td><p>Plot deconvolution results as a barplot</p></td>
      </tr><tr><td>
          <p><code><a href="plot_deconvolution.html">plot_deconvolution()</a></code> </p>
        </td>
        <td><p>Plot Deconvolution results</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

