<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ immunedeconv2</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunedeconv2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/immunedeconv2_vignette.html">immunedeconv2_vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/icbi-lab/immunedeconv2/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-main-package-methods" class="hasAnchor"><a href="#section-main-package-methods" class="anchor"></a>Main package methods</h2>
          <p class="section-desc"><p>Main methods to build signature matrices and deconvolute cell fractions.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="deconvolution_methods.html">deconvolution_methods</a></code> </p>
        </td>
        <td><p>List of supported immune deconvolution methods</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_model.html">build_model()</a></code> </p>
        </td>
        <td><p>Building the signature matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolute.html">deconvolute()</a></code> </p>
        </td>
        <td><p>Deconvolution</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-signature-matrix-creation-methods" class="hasAnchor"><a href="#section-signature-matrix-creation-methods" class="anchor"></a>Signature matrix creation methods</h2>
          <p class="section-desc"><p>Direct access to the individual signature matrix creation methods.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="build_model_momf.html">build_model_momf()</a></code> </p>
        </td>
        <td><p>Calculates the signature model with bisque</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_model_dwls.html">build_model_dwls()</a></code> </p>
        </td>
        <td><p>Signature matrix creation with DWLS using genes identified by DEAnalysisMAST()</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_model_scaden.html">build_model_scaden()</a></code> </p>
        </td>
        <td><p>Builds Scaden Model from scRNA data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_model_bisque.html">build_model_bisque()</a></code> </p>
        </td>
        <td><p>Calculates the signature model with bisque</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_model_cibersortx.html">build_model_cibersortx()</a></code> </p>
        </td>
        <td><p>Signature matrix creation with CibersortX</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-deconvolution-methods" class="hasAnchor"><a href="#section-deconvolution-methods" class="anchor"></a>Deconvolution methods</h2>
          <p class="section-desc"><p>Direct access to the individual deconvolution methods to access special features that are not available through <code>deconvolute</code>.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="deconvolute_momf.html">deconvolute_momf()</a></code> </p>
        </td>
        <td><p>Deconvolution Analysis using MOMF (via Nonnegative Factorization)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolute_dwls.html">deconvolute_dwls()</a></code> </p>
        </td>
        <td><p>Calculates the decomposition using the dwls algorithm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolute_scaden.html">deconvolute_scaden()</a></code> </p>
        </td>
        <td><p>Performs deconvolution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolute_bisque.html">deconvolute_bisque()</a></code> </p>
        </td>
        <td><p>Calculates the decomposition using the bisque algorithm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="deconvolute_cibersortx.html">deconvolute_cibersortx()</a></code> </p>
        </td>
        <td><p>Deconvolute with CibersortX</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-datasets" class="hasAnchor"><a href="#section-datasets" class="anchor"></a>Datasets</h2>
          <p class="section-desc"><p>List of datasets provided by the package.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="bulk.html">bulk</a></code> </p>
        </td>
        <td><p>Example RNA-seq dataset from Hoek et al.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="bulk_small.html">bulk_small</a></code> </p>
        </td>
        <td><p>Small example RNA-seq dataset from Hoek et al.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell_type_annotations.html">cell_type_annotations</a></code> </p>
        </td>
        <td><p>Cell type annotations for the example sigle-cell RNA-seq dataset from Maynard et al.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell_type_annotations_small.html">cell_type_annotations_small</a></code> </p>
        </td>
        <td><p>Cell type annotations for the small example sigle-cell RNA-seq dataset from Maynard et al.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="single_cell_data.html">single_cell_data</a></code> </p>
        </td>
        <td><p>Example sigle-cell RNA-seq dataset from Maynard et al.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="single_cell_data_small.html">single_cell_data_small</a></code> </p>
        </td>
        <td><p>Small example sigle-cell RNA-seq dataset from Maynard et al.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-helper" class="hasAnchor"><a href="#section-helper" class="anchor"></a>Helper</h2>
          <p class="section-desc"><p>Helper functions to extract repeating checks or work with non-R related tasks like Python or Docker.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="init_python.html">init_python()</a></code> </p>
        </td>
        <td><p>Initiates python environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="install_scaden.html">install_scaden()</a></code> </p>
        </td>
        <td><p>Install Scaden</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="python_available.html">python_available()</a></code> </p>
        </td>
        <td><p>Checks if python is available in environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_python.html">set_python()</a></code> </p>
        </td>
        <td><p>Defines python path.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_cibersortx_credentials.html">set_cibersortx_credentials()</a></code> </p>
        </td>
        <td><p>Setting the CibersortX Credentials</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="docker_available.html">docker_available()</a></code> </p>
        </td>
        <td><p>Docker availability check</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="docker_connectable.html">docker_connectable()</a></code> </p>
        </td>
        <td><p>Docker connectability check</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="verbose_wrapper.html">verbose_wrapper()</a></code> </p>
        </td>
        <td><p>A wrapper function whether to suppress messages</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-scaden" class="hasAnchor"><a href="#section-scaden" class="anchor"></a>Scaden</h2>
          <p class="section-desc"><p>Functions and objects related to the Scaden deconvolution_methods.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="scaden_checkload.html">scaden_checkload()</a></code> </p>
        </td>
        <td><p>Checks if scaden is installed.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaden_predict.html">scaden_predict()</a></code> </p>
        </td>
        <td><p>Predict cell proportions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaden_process.html">scaden_process()</a></code> </p>
        </td>
        <td><p>Pre-processes training data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaden_simulate.html">scaden_simulate()</a></code> </p>
        </td>
        <td><p>Simulates training data from scRNA data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaden_simulate_example.html">scaden_simulate_example()</a></code> </p>
        </td>
        <td><p>Simulates example Data provided by Scaden</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaden_train.html">scaden_train()</a></code> </p>
        </td>
        <td><p>Trains a Scaden model</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-anndata" class="hasAnchor"><a href="#section-anndata" class="anchor"></a>Anndata</h2>
          <p class="section-desc"><p>Functions related to anndata.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="anndata_check_load.html">anndata_check_load()</a></code> </p>
        </td>
        <td><p>Checks if anndata package is loaded</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build_anndata.html">build_anndata()</a></code> </p>
        </td>
        <td><p>Create anndata object from single cell expression matrix, cell type labels and gene symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_anndata.html">read_anndata()</a></code> </p>
        </td>
        <td><p>Read anndata object (.h5ad format)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_anndata.html">write_anndata()</a></code> </p>
        </td>
        <td><p>Write anndata object (.h5ad format)</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-conversion" class="hasAnchor"><a href="#section-data-conversion" class="anchor"></a>Data conversion</h2>
          <p class="section-desc"><p>Functions for data conversion.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="get_single_cell_expression_set.html">get_single_cell_expression_set()</a></code> </p>
        </td>
        <td><p>Generating a single cell expression set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="anndata_to_singlecellexperiment.html">anndata_to_singlecellexperiment()</a></code> </p>
        </td>
        <td><p>Convert AnnData to SingleCellExperiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="singlecellexperiment_to_anndata.html">singlecellexperiment_to_anndata()</a></code> </p>
        </td>
        <td><p>Convert SingleCellExperiment to AnnData object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrix_to_singlecellexperiment.html">matrix_to_singlecellexperiment()</a></code> </p>
        </td>
        <td><p>Convert count matrix to SingleCellExperiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="singlecellexperiment_to_matrix.html">singlecellexperiment_to_matrix()</a></code> </p>
        </td>
        <td><p>Convert SingleCellExperiment to Matrix + annotation vector for celltypes</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


