<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>immunedeconv2_vignette • immunedeconv2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="immunedeconv2_vignette">
<meta property="og:description" content="immunedeconv2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">immunedeconv2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/immunedeconv2_vignette.html">immunedeconv2_vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PelzKo/immunedeconv2/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="immunedeconv2_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>immunedeconv2_vignette</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PelzKo/immunedeconv2/blob/master/vignettes/immunedeconv2_vignette.Rmd"><code>vignettes/immunedeconv2_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>immunedeconv2_vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>1. Introduction</h1>
<p>Cell-type heterogeneity is a strong confounder as well as source of information in bulk RNA sequencing (RNA-seq) data. Computational methods for the deconvolution of bulk RNA-seq data can infer the cellular composition of heterogeneous samples by leveraging precomputed expression signatures that represent the transcriptional profiles of the cell types of interest. Second-generation deconvolution tools allow deriving cell-type-specific expression signatures from single-cell RNA-seq (scRNA-seq) enabling, in principle, the deconvolution of any cell type profiled with single-cell technologies.</p>
<p>Our interface unifies multiple second-generation deconvolution methods to facilitate the usage of these tools and clear the way for future benchmarking of these methods.</p>
</div>
<div id="deconvolution" class="section level1">
<h1 class="hasAnchor">
<a href="#deconvolution" class="anchor"></a>2. Deconvolution</h1>
<p>In this section, we want to give an overview about the different methods and provide an example of how to use them.</p>
<div id="installation-and-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-and-setup" class="anchor"></a>Installation and Setup</h2>
<p>Install the CRAN package devtools and use it to install immunedeconv2 from <a href="https://github.com/">GitHub</a>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"PelzKo/immunedeconv2"</span><span class="op">)</span></code></pre></div>
</div>
<div id="input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input Data</h2>
<p>First of all, you need some data to run the deconvolution with.</p>
<ul>
<li>Single cell RNA-seq data
<ul>
<li>
<em>Genes</em> x <em>Cells</em> matrix</li>
<li>Counts are <em>TPM normalized</em>, <em>not log-transformed</em>
</li>
<li>Rownames (genenames) are provided in the same format as in the bulk RNA-seq data, for instance HGNC symbols</li>
</ul>
</li>
<li>Cell type annotations
<ul>
<li>Vector containing cell type annotations</li>
<li>Annotations are in the same order as the columns of the single cell matrix</li>
</ul>
</li>
<li>Bulk RNA-seq data
<ul>
<li>
<em>Genes</em> x <em>Samples</em> matrix</li>
<li>Rownames (genenames) are provided in the same format as in the sc RNA-seq data, for instance HGNC symbols</li>
</ul>
</li>
</ul>
<p>We also provide some sample data, a bigger and a smaller set of single cell RNA-seq matrix (Maynard et al. 2020), cell type annotations and bulk RNA-seq matrix (Hoek et al. 2015). You can simply load it with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"single_cell_data"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"cell_type_annotations"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bulk"</span><span class="op">)</span></code></pre></div>
<p>For the small dataset, append "_small" to the names, e.g. </p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"single_cell_data_small"</span><span class="op">)</span></code></pre></div>
</div>
<div id="signature-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#signature-matrix" class="anchor"></a>Signature Matrix</h2>
<p>To calculate a cell-type-specific expression signature, you can use the build_model function using the single cell RNA-seq matrix and cell type annotations specified above. The parameter method can be set to one of the four methods listed below. More information of the methods is provided in section 3.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_data</span>, <span class="va">cell_type_annotations</span>, <span class="va">method</span><span class="op">)</span></code></pre></div>
<ul>
<li>Bisque (“bisque”)</li>
<li>DWLS (“dwls”)</li>
<li>MOMF (“momf”)</li>
<li>Scaden (“scaden”)</li>
</ul>
<p>Note: MOMF and Scaden also need the bulk RNA-seq data to calculate a signature matrix. Please provide this parameter as well when running build_model with one of these two methods.</p>
<p>Additional parameters for the different methods can be included in the method call as well. For further information on which options you have, see the documentation of each tool.</p>
</div>
<div id="deconvolution-of-bulk-rna-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#deconvolution-of-bulk-rna-seq-data" class="anchor"></a>Deconvolution of Bulk RNA-seq Data</h2>
<p>After building the specific signature matrix, you can calculate the cell properties in bulk RNA-seq data with the deconvolute function. For this, you need to provide the bulk data, signature matrix and you can, again, choose between the four methods listed in the previous section.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk</span>, <span class="va">signature_matrix</span>, <span class="va">method</span><span class="op">)</span></code></pre></div>
<p>Note: Bisque and MOMF additionally need the single cell RNA-seq matrix, that you already used to build the signature matrix, for their deconvolution. As a fifth parameter, Bisque also needs cell type annotations, which were also already needed for building the model.</p>
<p>As for the build_model function, specific parameters can be passed on to the deconvolution methods.</p>
</div>
<div id="example-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#example-workflow" class="anchor"></a>Example Workflow</h2>
<p>This workflow provides a more detailed example of how to run the deconvolution with Bisque.</p>
<p>First of all, make sure you load all the libraries we need for this analysis. If the packages are not available, install them via install_packages().</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PelzKo/immunedeconv2">immunedeconv2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<p>To give you a feeling for the data, this is what the first few rows and columns of our small single cell RNA-seq dataset look like:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">single_cell_data_small</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">SRR10779801</th>
<th align="right">SRR10797955</th>
<th align="right">SRR10779362</th>
<th align="right">SRR10788422</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">TSPAN6</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">FGR</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">4.597211</td>
</tr>
<tr class="odd">
<td align="left">CYP51A1</td>
<td align="right">45.42429</td>
<td align="right">2.919613</td>
<td align="right">0.279393</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">AOC1</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">WNT16</td>
<td align="right">0.00000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
</tr>
</tbody>
</table>
<p>Now we build the signature matrix with Bisque and look at the values for the first few genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">signatureMatrix</span> <span class="op">&lt;-</span> <span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_data_small</span>, <span class="va">cell_type_annotations_small</span>, <span class="st">"bisque"</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">signatureMatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">B cell</th>
<th align="right">Monocyte conventional</th>
<th align="right">NK cell</th>
<th align="right">T cell CD4</th>
<th align="right">T cell CD8</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">TSPAN6</td>
<td align="right">0.0000000</td>
<td align="right">0.00000</td>
<td align="right">0.000</td>
<td align="right">0.0000000</td>
<td align="right">1.190598</td>
</tr>
<tr class="even">
<td align="left">FGR</td>
<td align="right">0.1098711</td>
<td align="right">79.15949</td>
<td align="right">37.762</td>
<td align="right">0.1575649</td>
<td align="right">28.259981</td>
</tr>
<tr class="odd">
<td align="left">CYP51A1</td>
<td align="right">51.5579060</td>
<td align="right">44.05069</td>
<td align="right">0.000</td>
<td align="right">7.8737725</td>
<td align="right">40.540089</td>
</tr>
<tr class="even">
<td align="left">AOC1</td>
<td align="right">0.0000000</td>
<td align="right">0.00000</td>
<td align="right">0.000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">WNT16</td>
<td align="right">0.0000000</td>
<td align="right">0.00000</td>
<td align="right">0.000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
</tr>
</tbody>
</table>
<p>We can use this signature matrix to run the deconvolution with Bisque.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk_small</span>, <span class="va">signatureMatrix</span>, <span class="st">"bisque"</span>, <span class="va">single_cell_data_small</span>, <span class="va">cell_type_annotations_small</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">B cell</th>
<th align="right">Monocyte conventional</th>
<th align="right">NK cell</th>
<th align="right">T cell CD4</th>
<th align="right">T cell CD8</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HD30_PBMC_0</td>
<td align="right">0.0000000</td>
<td align="right">0.2649449</td>
<td align="right">0.4888424</td>
<td align="right">0.2462127</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_1</td>
<td align="right">0.0116098</td>
<td align="right">0.7700322</td>
<td align="right">0.2183581</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">HD30_PBMC_3</td>
<td align="right">0.0000000</td>
<td align="right">0.0780721</td>
<td align="right">0.2319889</td>
<td align="right">0.3724884</td>
<td align="right">0.3174506</td>
</tr>
<tr class="even">
<td align="left">HD30_PBMC_7</td>
<td align="right">0.3936133</td>
<td align="right">0.0000000</td>
<td align="right">0.2048515</td>
<td align="right">0.2541864</td>
<td align="right">0.1473488</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_0</td>
<td align="right">0.2688500</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.7311500</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_1</td>
<td align="right">0.0444611</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.9555389</td>
</tr>
<tr class="odd">
<td align="left">HD31_PBMC_3</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">1.0000000</td>
</tr>
<tr class="even">
<td align="left">HD31_PBMC_7</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">1.0000000</td>
</tr>
</tbody>
</table>
</div>
<div id="plot-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-results" class="anchor"></a>Plot Results</h2>
<p>Here are some further ideas of how you could plot your deconvolution results.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span><span class="va">bulk_small</span>, <span class="va">signatureMatrix</span>, <span class="st">"bisque"</span>, <span class="va">single_cell_data_small</span>, <span class="va">cell_type_annotations_small</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Var2</span>, x<span class="op">=</span><span class="va">value</span>, fill<span class="op">=</span><span class="va">Var1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"stack"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Bisque"</span>, y<span class="op">=</span><span class="st">"sample"</span>, x<span class="op">=</span><span class="st">"proportion"</span>, fill<span class="op">=</span><span class="st">"cell type"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Accent"</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"B cell"</span>,<span class="st">"Monocyte conventional"</span>,<span class="st">"NK cell"</span>,<span class="st">"T cell CD4"</span>,<span class="st">"T cell CD8"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="immunedeconv2_vignette_files/figure-html/plot-1.png" width="672"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Var2</span>, x<span class="op">=</span><span class="va">value</span>, color<span class="op">=</span><span class="va">Var1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">Var1</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Bisque"</span>, y<span class="op">=</span><span class="st">"sample"</span>, x<span class="op">=</span><span class="st">"proportion"</span>, color<span class="op">=</span><span class="st">"cell type"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Accent"</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"B cell"</span>,<span class="st">"Monocyte conventional"</span>,<span class="st">"NK cell"</span>,<span class="st">"T cell CD4"</span>,<span class="st">"T cell CD8"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="immunedeconv2_vignette_files/figure-html/plot-2.png" width="672"> In the next figures you can see, how the other methods are used and how the cell proportions would be distributed based on their calculations. Here are some further ideas of how you could plot your deconvolution results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/build_model.html">build_model</a></span><span class="op">(</span><span class="va">single_cell_data_small</span>, <span class="va">cell_type_annotations_small</span>, <span class="st">"momf"</span>, <span class="va">bulk_small</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu">immunedeconv2</span><span class="fu">::</span><span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span>bulk_gene_expression <span class="op">=</span> <span class="va">bulk_small</span>, method <span class="op">=</span> <span class="st">"momf"</span>, single_cell_object <span class="op">=</span> <span class="va">single_cell_data_small</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">Var1</span>, x<span class="op">=</span><span class="va">value</span>, fill<span class="op">=</span><span class="va">Var2</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"stack"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"MOMF"</span>, y<span class="op">=</span><span class="st">"sample"</span>, x<span class="op">=</span><span class="st">"proportion"</span>, fill<span class="op">=</span><span class="st">"cell type"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Accent"</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"B cell"</span>,<span class="st">"Monocyte conventional"</span>,<span class="st">"NK cell"</span>,<span class="st">"T cell CD4"</span>,<span class="st">"T cell CD8"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="immunedeconv2_vignette_files/figure-html/otherMethods-1.png" width="672"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># immunedeconv2::build_model(single_cell_object = single_cell_data_small, cell_type_annotations = cell_type_annotations_small, method = "scaden", bulk_gene_expression = bulk_small) %&gt;%</span>
<span class="co"># immunedeconv2::deconvolute(bulk_gene_expression = bulk_small, method = "scaden", single_cell_object = single_cell_data_small) %&gt;%</span>
<span class="co"># reshape2::melt() %&gt;%</span>
<span class="co"># ggplot(aes(y=Var1, x=value, fill=Var2))+geom_bar(stat="identity", position = "stack")+labs(title="Scaden", y="sample", x="proportion", fill="cell type")+scale_fill_brewer(palette="Accent", breaks=c("B cell","Monocyte conventional","NK cell","T cell CD4","T cell CD8"))</span></code></pre></div>
</div>
</div>
<div id="methods" class="section level1">
<h1 class="hasAnchor">
<a href="#methods" class="anchor"></a>3. Methods</h1>
<p>In this section, we provide a short overview over the deconvolution tools that can be used via this package.</p>
<div id="bisque" class="section level2">
<h2 class="hasAnchor">
<a href="#bisque" class="anchor"></a>Bisque</h2>
<p>Bisque(Jew et al., 2020) takes advantage of matched bulk and scRNA-seq samples to improve accuracy of cell-type estimates. Using linear regression, it corrects for platform-specific biases between pseudo-bulk reference profiles derived from scRNA-seq and bulk RNA-seq samples. It then applies NNLS-regression to deconvolute other bulk RNA-seq samples.</p>
<p>Citation: Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., &amp; Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. <a href="https://doi.org/10.1038/s41467-020-16607-9" class="uri">https://doi.org/10.1038/s41467-020-16607-9</a></p>
</div>
<div id="dwls" class="section level2">
<h2 class="hasAnchor">
<a href="#dwls" class="anchor"></a>DWLS</h2>
<p>Tsoucas et al(Tsoucas et al., 2019) introduce dampened weighted least squares (DWLS)-regression, which improves over ordinary least squares regression or 𝜈-SVR in that it attributes higher weight to rare cell types and marker genes with a low average expression level. They show that their method achieves favorable performance when estimating the abundance of infrequent cell-types.</p>
<p>Citation: Tsoucas, D., Dong, R., Chen, H., Zhu, Q., Guo, G., &amp; Yuan, G.-C. (2019). Accurate estimation of cell-type composition from gene expression data. Nature Communications, 10(1), 2975. <a href="https://doi.org/10.1038/s41467-019-10802-z" class="uri">https://doi.org/10.1038/s41467-019-10802-z</a></p>
</div>
<div id="momf" class="section level2">
<h2 class="hasAnchor">
<a href="#momf" class="anchor"></a>MOMF</h2>
<p>With MOMF (Sun et al., 2019), Multi-Omics Matrix Factorization, models the cell type specific mean in sc RNA-seq data by also accounting for the uncertainty of these cell type specific mean gene expression levels. Alternating Direction Method of Multipliers algorithm is then used to estimate the parameters in bulk RNA-seq downstream analyses.</p>
<p>Citation: Xifang Sun, Shiquan Sun, and Sheng Yang. An efficient and flexible method for deconvoluting bulk RNAseq data with single-cell RNAseq data, 2019, DIO: 10.5281/zenodo.3373980</p>
</div>
<div id="scaden" class="section level2">
<h2 class="hasAnchor">
<a href="#scaden" class="anchor"></a>Scaden</h2>
<p>Scaden(Menden et al., n.d.) leverages a deep neural network (DNN) for estimating cell-type proportions. Instead of explicitly building a signature matrix, the DNN implicitly learns which features are important for a certain cell type. The DNN is trained by simulating bulk RNA-seq samples with known cell-type proportions from scRNA-seq datasets. To increase robustness, the training process allows to flexibly integrate multiple scRNA-seq datasets and, optionally, bulk RNA-seq samples with “gold standard” measurements such as FACS.</p>
<p>Citation: Menden, K., Marouf, M., Oller, S., Dalmia, A., Kloiber, K., Heutink, P., &amp; Bonn, S. (n.d.). Deep-learning-based cell composition analysis from tissue expression profiles. <a href="https://doi.org/10.1101/659227" class="uri">https://doi.org/10.1101/659227</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Konstantin Pelz, Nicolas Goedert, Yuyu Liang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
