<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Deconvolution of the tumor microenvironment with omnideconv • omnideconv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Deconvolution of the tumor microenvironment with omnideconv">
<meta property="og:description" content="omnideconv">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">omnideconv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/omnideconv_example.html">Deconvolution of the tumor microenvironment with omnideconv</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/omnideconv/omnideconv/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Deconvolution of the tumor microenvironment with omnideconv</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/omnideconv/omnideconv/blob/HEAD/vignettes/omnideconv_example.Rmd" class="external-link"><code>vignettes/omnideconv_example.Rmd</code></a></small>
      <div class="hidden name"><code>omnideconv_example.Rmd</code></div>

    </div>

    
    
<pre><code><span><span class="co">#&gt; Error in knitr::include_graphics("logo/omnideconv_logo.jpg"): Cannot find the file(s): "logo/omnideconv_logo.jpg"</span></span></code></pre>
<div class="section level2">
<h2 id="introduction-and-dataset">1. Introduction and dataset<a class="anchor" aria-label="anchor" href="#introduction-and-dataset"></a>
</h2>
<p>In this vignette, we will use the omnideconv package to deconvolve a bulk RNA-seq dataset from 24 breast cancer patients with two different methods (DWLS and BayesPrism). The datasets are from a recent breast cancer study <span class="citation">(Wu et al. <a href="#ref-Wu2021" role="doc-biblioref">2021</a>)</span>. This study provides access to a primary-tumor single cell RNA-seq (scRNA-seq) dataset from 26 breast cancer patients across three major cancer subtypes (ER+, HER2+, TNBC). The dataset includes cell-type annotation for three resolution levels. In addition, the data includes bulk RNA-seq sequencing for 24 of the patients. In this chapter, we use multi-sample, breast-cancer scRNA-seq atlas (100,064 cells) as a reference to train the methods for the deconvolution of the bulk RNA-seq samples. The single-cell and bulk RNA-seq data is deposited on GEO, under accession number: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE176078" class="external-link">GSE176078</a>. The single cell data comes with the author’s cell type annotations.<br>
We will need to download and unzip the datasets (<code>GSE176078_Wu_etal_2021_BRCA_scRNASeq.tar.gz</code>, <code>GSE176078_Wu_etal_2021_bulkRNAseq_raw_counts.txt.gz</code>), and store them in the working directory. For this example analysis, we will also need to retrieve the additional clinical information about the patients – although it is not required by omnideconv. This is available in the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9044823/bin/NIHMS1793222-supplement-1793222_Sup_Tab_1-11.xlsx" class="external-link">Supplementary Table 1</a>, included with the paper supplementary materials</p>
</div>
<div class="section level2">
<h2 id="library-loading">2. Library loading<a class="anchor" aria-label="anchor" href="#library-loading"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/omnideconv/omnideconv" class="external-link">omnideconv</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Error in gzfile(file, "rb"): cannot open the connection</span></span>
<span><span class="co">#&gt; Error in gzfile(file, "rb"): cannot open the connection</span></span>
<span><span class="co">#&gt; Error in gzfile(file, "rb"): cannot open the connection</span></span>
<span><span class="co">#&gt; Error in gzfile(file, "rb"): cannot open the connection</span></span>
<span><span class="co">#&gt; Error in gzfile(file, "rb"): cannot open the connection</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="single-cell-data-processing">3. Single cell data processing<a class="anchor" aria-label="anchor" href="#single-cell-data-processing"></a>
</h2>
<p>Although not strictly required by omnideconv, we suggest performing quality control and filtering of the input scRNA-seq data according to the best practice <span class="citation">(Heumos et al. <a href="#ref-Heumos2023" role="doc-biblioref">2023</a>)</span>, to ensure the best training conditions for deconvolution algorithms. We first pre-process the single-cell dataset to remove low-quality cells. We will use the R package <a href="https://github.com/satijalab/seurat" class="external-link">Seurat</a> <span class="citation">(Hao et al. <a href="#ref-Hao2023" role="doc-biblioref">2023</a>)</span>, which allows fast and easy manipulation of single-cell data. We will create a Seurat object with the cell counts and their metadata of interest curated by the authors, which include cell-type annotation on three levels of resolution:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single.cell.data</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span></span>
<span>  mtx <span class="op">=</span> <span class="st">'C:/Users/c1041161/book_chapter/Wu_etal_2021_BRCA_scRNASeq/count_matrix_sparse.mtx'</span>,</span>
<span>  cells <span class="op">=</span> <span class="st">'C:/Users/c1041161/book_chapter/Wu_etal_2021_BRCA_scRNASeq/count_matrix_barcodes.tsv'</span>,</span>
<span>  features <span class="op">=</span> <span class="st">'C:/Users/c1041161/book_chapter/Wu_etal_2021_BRCA_scRNASeq/count_matrix_genes.tsv'</span>,</span>
<span>  feature.column <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">single.cell.metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">'C:/Users/c1041161/book_chapter/Wu_etal_2021_BRCA_scRNASeq/metadata.csv'</span>, </span>
<span>                                   sep <span class="op">=</span> <span class="st">','</span>,</span>
<span>                                   header <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                   row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single.cell.data</span> <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">single.cell.data</span>, </span>
<span>                                      project<span class="op">=</span><span class="st">'Wu_dataset'</span>, </span>
<span>                                      assay<span class="op">=</span><span class="st">'RNA'</span>, </span>
<span>                                      min.cells <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                      min.features <span class="op">=</span> <span class="fl">1</span>, meta.data <span class="op">=</span> <span class="va">single.cell.metadata</span><span class="op">)</span></span></code></pre></div>
<p>We can have an overview of the number of cells per cell type in the dataset:</p>
<table class="table">
<caption>Number of cells per cell type</caption>
<thead><tr class="header">
<th align="left">celltype_major</th>
<th align="right">number_cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B-cells</td>
<td align="right">3206</td>
</tr>
<tr class="even">
<td align="left">CAFs</td>
<td align="right">6573</td>
</tr>
<tr class="odd">
<td align="left">Cancer Epithelial</td>
<td align="right">24489</td>
</tr>
<tr class="even">
<td align="left">Endothelial</td>
<td align="right">7605</td>
</tr>
<tr class="odd">
<td align="left">Myeloid</td>
<td align="right">9675</td>
</tr>
<tr class="even">
<td align="left">Normal Epithelial</td>
<td align="right">4355</td>
</tr>
<tr class="odd">
<td align="left">Plasmablasts</td>
<td align="right">3524</td>
</tr>
<tr class="even">
<td align="left">PVL</td>
<td align="right">5423</td>
</tr>
<tr class="odd">
<td align="left">T-cells</td>
<td align="right">35214</td>
</tr>
</tbody>
</table>
<p>In order to remove low quality cells, we will follow the <a href="https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html" class="external-link">best practices for single cell normalization</a> <span class="citation">(Heumos et al. <a href="#ref-Heumos2023" role="doc-biblioref">2023</a>)</span>. We will perform quality control on each cell by considering metrics such as the number of total counts, the number of expressed features (genes), and the fraction of mitochondrial genes per cell. We will remove cells that have <span class="math inline">\(MAD = median(|X_i - median(X)|)\)</span>, where</p>
<ul>
<li>
<span class="math inline">\(X_i\)</span> is the respective metric of an observation <span class="math inline">\(i\)</span> (cell), for example the number of detected genes<br>
</li>
<li>
<span class="math inline">\(MAD\)</span> is the Mead Absolute Deviation, computed as <span class="math inline">\(MAD = median(|X_i - median(X)|)\)</span><br>
</li>
<li>
<span class="math inline">\(n\)</span> is set to 5 for the number of counts and expressed features, and to 3 for the fraction of mitochondrial genes</li>
</ul>
<p>In order to do this filtering, we will create a function to identify the outliers for each metric</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_outlier</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">SeuratObject</span>, <span class="va">metric</span>, <span class="va">nmads</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/parse.html" class="external-link">parse</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"M &lt;- SeuratObject$"</span>,<span class="va">metric</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">outlier</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">M</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op">-</span> <span class="va">nmads</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/mad.html" class="external-link">mad</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">M</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op">+</span> <span class="va">nmads</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/mad.html" class="external-link">mad</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">outlier</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">check_outliers_nFeature</span> <span class="op">&lt;-</span> <span class="fu">is_outlier</span><span class="op">(</span><span class="va">single.cell.data</span>, <span class="st">'nFeature_RNA'</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">check_outliers_nCount</span> <span class="op">&lt;-</span> <span class="fu">is_outlier</span><span class="op">(</span><span class="va">single.cell.data</span>, <span class="st">'nCount_RNA'</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">check_outliers_mito</span> <span class="op">&lt;-</span> <span class="fu">is_outlier</span><span class="op">(</span><span class="va">single.cell.data</span>, <span class="st">'percent.mito'</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">non_outliers_nFeature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">check_outliers_nFeature</span><span class="op">[</span><span class="op">!</span><span class="va">check_outliers_nFeature</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">non_outliers_nCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">check_outliers_nCount</span><span class="op">[</span><span class="op">!</span><span class="va">check_outliers_nCount</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">non_outliers_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">check_outliers_mito</span><span class="op">[</span><span class="op">!</span><span class="va">check_outliers_mito</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>We will retain only those that satisfy all three of the conditions described above.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">non_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">non_outliers_nFeature</span>, <span class="va">non_outliers_nCount</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">non_outliers_mito</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single.cell.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">single.cell.data</span>, cells <span class="op">=</span> <span class="va">non_outliers</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">single.cell.data</span><span class="op">$</span><span class="va">celltype_major</span>, dnn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"celltype_major"</span><span class="op">)</span><span class="op">)</span>, responseName <span class="op">=</span> <span class="st">"number_cells"</span><span class="op">)</span></span>
<span><span class="co">#&gt;      celltype_major number_cells</span></span>
<span><span class="co">#&gt; 1           B-cells         3150</span></span>
<span><span class="co">#&gt; 2              CAFs         6154</span></span>
<span><span class="co">#&gt; 3 Cancer Epithelial        16613</span></span>
<span><span class="co">#&gt; 4       Endothelial         6991</span></span>
<span><span class="co">#&gt; 5           Myeloid         8855</span></span>
<span><span class="co">#&gt; 6 Normal Epithelial         3533</span></span>
<span><span class="co">#&gt; 7      Plasmablasts         3164</span></span>
<span><span class="co">#&gt; 8               PVL         5120</span></span>
<span><span class="co">#&gt; 9           T-cells        34991</span></span></code></pre></div>
<table class="table">
<caption>Number of cells per cell type after quality filtering</caption>
<thead><tr class="header">
<th align="left">celltype_major</th>
<th align="right">number_cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B-cells</td>
<td align="right">3150</td>
</tr>
<tr class="even">
<td align="left">CAFs</td>
<td align="right">6154</td>
</tr>
<tr class="odd">
<td align="left">Cancer Epithelial</td>
<td align="right">16613</td>
</tr>
<tr class="even">
<td align="left">Endothelial</td>
<td align="right">6991</td>
</tr>
<tr class="odd">
<td align="left">Myeloid</td>
<td align="right">8855</td>
</tr>
<tr class="even">
<td align="left">Normal Epithelial</td>
<td align="right">3533</td>
</tr>
<tr class="odd">
<td align="left">Plasmablasts</td>
<td align="right">3164</td>
</tr>
<tr class="even">
<td align="left">PVL</td>
<td align="right">5120</td>
</tr>
<tr class="odd">
<td align="left">T-cells</td>
<td align="right">34991</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="bulk-data-preprocessing">4. Bulk data preprocessing<a class="anchor" aria-label="anchor" href="#bulk-data-preprocessing"></a>
</h2>
<p>We will now read in the bulk sequencing data file, which consists of 24 samples.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bulk.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">'C:/Users/c1041161/book_chapter/GSE176078_Wu_etal_2021_bulkRNAseq_raw_counts.txt'</span>, skip<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">header</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">'C:/Users/c1041161/book_chapter/GSE176078_Wu_etal_2021_bulkRNAseq_raw_counts.txt'</span>, </span>
<span>                     header <span class="op">=</span> <span class="cn">FALSE</span>, nrows <span class="op">=</span> <span class="fl">1</span>, skipNul <span class="op">=</span> <span class="cn">TRUE</span>, sep<span class="op">=</span><span class="st">'\t'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bulk.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Genes'</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'A|N'</span>, <span class="st">''</span>, <span class="va">header</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bulk.data</span> <span class="op">&lt;-</span> <span class="va">bulk.data</span><span class="op">[</span><span class="va">bulk.data</span><span class="op">$</span><span class="va">Genes</span> <span class="op">!=</span> <span class="st">''</span>, <span class="op">]</span></span>
<span><span class="va">bulk.data</span> <span class="op">&lt;-</span> <span class="fu">column_to_rownames</span><span class="op">(</span><span class="va">bulk.data</span>, <span class="st">'Genes'</span><span class="op">)</span></span>
<span><span class="va">bulk.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">bulk.data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="subsampling-of-single-cell-data">5. Subsampling of single cell data<a class="anchor" aria-label="anchor" href="#subsampling-of-single-cell-data"></a>
</h2>
<p>The various methods included in omnideconv rely on the single cell dataset that will be used to train them for the deconvolution of those specific cell types. This training involves the optimization of internal features of the methods and can happen in different ways. Some methods use the single cell data to build a ‘signature matrix’, i.e. a reduced transcriptional fingerprints of the cell types provided, while others use this data in a statistical or deep learning model. Since single cell datasets can often encompass thousands of cells, we will need to subsample it in order to be able to run the analysis on our machines. In this case we will retain a maximum of 200 cells per cell type, but this step can be costumed, or eventually skipped, depending on the computational resources available.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_cells_per_celltype</span> <span class="op">=</span> <span class="fl">200</span></span>
<span></span>
<span><span class="va">sampled.metadata</span> <span class="op">&lt;-</span> <span class="va">single.cell.data</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">rownames_to_column</span><span class="op">(</span><span class="va">.</span>, <span class="st">'barcode'</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">group_by</span><span class="op">(</span><span class="va">.</span>, <span class="va">celltype_major</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">nest</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>            </span>
<span>      <span class="fu">mutate</span><span class="op">(</span>n <span class="op">=</span>  <span class="fu">map_dbl</span><span class="op">(</span><span class="va">data</span>, <span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">mutate</span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">n</span>, <span class="va">max_cells_per_celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">mutate</span><span class="op">(</span>samp <span class="op">=</span> <span class="fu">map2</span><span class="op">(</span><span class="va">data</span>, <span class="va">n</span>, <span class="va">sample_n</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">unnest</span><span class="op">(</span><span class="va">samp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single.cell.data.sampled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">single.cell.data</span>, cells <span class="op">=</span> <span class="va">sampled.metadata</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#as.data.frame(table(single.cell.data.sampled$celltype_major, dnn = list("celltype_major")), responseName = "number_cells")</span></span></code></pre></div>
<table class="table">
<caption>Number of cells per cell type after subsampling</caption>
<thead><tr class="header">
<th align="left">celltype_major</th>
<th align="right">number_cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B-cells</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">CAFs</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Cancer Epithelial</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Endothelial</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Myeloid</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Normal Epithelial</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Plasmablasts</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">PVL</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">T-cells</td>
<td align="right">200</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="deconvolution-of-the-bulk-data">6. Deconvolution of the bulk data<a class="anchor" aria-label="anchor" href="#deconvolution-of-the-bulk-data"></a>
</h2>
<p>Each methods has different requirements, but in general to compute the deconvolution results we will need the single cell counts matrix, the cell type annotations and the information on the individual/experiment fom which the cells were retrieved (batch ID).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">single.cell.data.sampled</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">cell.type.annotations</span> <span class="op">&lt;-</span> <span class="va">single.cell.data.sampled</span><span class="op">$</span><span class="va">celltype_major</span></span>
<span><span class="va">batch.ids</span> <span class="op">&lt;-</span> <span class="va">single.cell.data.sampled</span><span class="op">$</span><span class="va">orig.ident</span></span></code></pre></div>
<div class="section level3">
<h3 id="deconvolution-with-dwls">6.1 Deconvolution with DWLS<a class="anchor" aria-label="anchor" href="#deconvolution-with-dwls"></a>
</h3>
<p>Now we’re going to deconvolute the bulk dataset with different methods. The first one we are going to use is called <a href="https://github.com/dtsoucas/DWLS" class="external-link">DWLS</a> <span class="citation">(Tsoucas et al. <a href="#ref-Tsoucas2019" role="doc-biblioref">2019</a>)</span> and performs the deconvolution in a two-steps process. First, the single cell data is used to build a signature matrix using the <code>omnideconv</code> function <code>build_model</code>. DWLS looks for differentially expressed genes that discriminate across cell types, and can do so with two approaches based either on the Seurat <span class="citation">(Hao et al. <a href="#ref-Hao2023" role="doc-biblioref">2023</a>)</span> “bimod” test <span class="citation">(McDavid et al. <a href="#ref-McDavid2012" role="doc-biblioref">2012</a>)</span> or on MAST <span class="citation">(Finak et al. <a href="#ref-Finak2015" role="doc-biblioref">2015</a>)</span>. MAST improves the former model, but has an increased computational requirement <span class="citation">(Nault et al. <a href="#ref-Nault2022" role="doc-biblioref">2022</a>)</span>. The authors recommend using this method on the smaller datasets, and to switch to Seurat if the analysis with MAST cannot be completed. To reduce MAST’s computational time, we introduced a second version of the MAST-based function (mast_optimized) that speeds up the process compared to the original implementation:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We need to insert the normalization as well</span></span>
<span><span class="va">signature.matrix.dwls</span> <span class="op">&lt;-</span> <span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="../reference/build_model.html">build_model</a></span><span class="op">(</span>single_cell_object <span class="op">=</span> <span class="va">counts.matrix</span>,</span>
<span>                        cell_type_annotations <span class="op">=</span> <span class="va">cell.type.annotations</span>,</span>
<span>                        method <span class="op">=</span> <span class="st">'dwls'</span>, </span>
<span>                        dwls_method <span class="op">=</span> <span class="st">'mast_optimized'</span><span class="op">)</span></span></code></pre></div>
<p>This signature is optimized so that the genes selected are the ones that help to discriminate across cell types.</p>
<p>The signature is now used for the deconvolution of the bulk RNAseq, which is performed with the <code>omnideconv</code> function <code>deconvolute</code>. DWLS computes the cell fractions performing one of Ordinary Least Squares (OLS) Regression, Support Vector Regression (SVR) or the Dampened Weighted Least Squares Regression (DWLS) that was introduced with the method. This last regression method is shown to outperform the others when it comes to the detection of rare cell types:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deconvolution.results.dwls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span>bulk_gene_expression <span class="op">=</span> <span class="va">bulk.data</span>,</span>
<span>                                          signature <span class="op">=</span> <span class="va">signature.matrix.dwls</span>,</span>
<span>                                          method<span class="op">=</span><span class="st">'dwls'</span>,</span>
<span>                                          dwls_submethod <span class="op">=</span> <span class="st">'DampenedWLS'</span><span class="op">)</span></span></code></pre></div>
<p>We will now obtain, for every sample, a set of cell type fractions for each cell type that was included in the provided single cell dataset.</p>
<table style="width:100%;" class="table">
<caption>Cell type fractions obtained by DWLS</caption>
<colgroup>
<col width="8%">
<col width="7%">
<col width="5%">
<col width="16%">
<col width="11%">
<col width="7%">
<col width="16%">
<col width="12%">
<col width="5%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">B-cells</th>
<th align="right">CAFs</th>
<th align="right">Cancer Epithelial</th>
<th align="right">Endothelial</th>
<th align="right">Myeloid</th>
<th align="right">Normal Epithelial</th>
<th align="right">Plasmablasts</th>
<th align="right">PVL</th>
<th align="right">T-cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">CID3586</td>
<td align="right">0.231</td>
<td align="right">0.238</td>
<td align="right">0.170</td>
<td align="right">0.083</td>
<td align="right">0.005</td>
<td align="right">0.035</td>
<td align="right">0.000</td>
<td align="right">0.022</td>
<td align="right">0.215</td>
</tr>
<tr class="even">
<td align="left">CID3921</td>
<td align="right">0.026</td>
<td align="right">0.177</td>
<td align="right">0.145</td>
<td align="right">0.114</td>
<td align="right">0.019</td>
<td align="right">0.068</td>
<td align="right">0.004</td>
<td align="right">0.000</td>
<td align="right">0.447</td>
</tr>
<tr class="odd">
<td align="left">CID3941</td>
<td align="right">0.000</td>
<td align="right">0.267</td>
<td align="right">0.373</td>
<td align="right">0.091</td>
<td align="right">0.002</td>
<td align="right">0.180</td>
<td align="right">0.000</td>
<td align="right">0.030</td>
<td align="right">0.059</td>
</tr>
<tr class="even">
<td align="left">CID3948</td>
<td align="right">0.466</td>
<td align="right">0.107</td>
<td align="right">0.264</td>
<td align="right">0.049</td>
<td align="right">0.007</td>
<td align="right">0.000</td>
<td align="right">0.001</td>
<td align="right">0.020</td>
<td align="right">0.086</td>
</tr>
<tr class="odd">
<td align="left">CID3963</td>
<td align="right">0.112</td>
<td align="right">0.208</td>
<td align="right">0.000</td>
<td align="right">0.088</td>
<td align="right">0.103</td>
<td align="right">0.345</td>
<td align="right">0.000</td>
<td align="right">0.020</td>
<td align="right">0.123</td>
</tr>
<tr class="even">
<td align="left">CID4066</td>
<td align="right">0.049</td>
<td align="right">0.342</td>
<td align="right">0.174</td>
<td align="right">0.102</td>
<td align="right">0.019</td>
<td align="right">0.155</td>
<td align="right">0.000</td>
<td align="right">0.049</td>
<td align="right">0.111</td>
</tr>
<tr class="odd">
<td align="left">CID4067</td>
<td align="right">0.000</td>
<td align="right">0.338</td>
<td align="right">0.477</td>
<td align="right">0.084</td>
<td align="right">0.007</td>
<td align="right">0.018</td>
<td align="right">0.000</td>
<td align="right">0.006</td>
<td align="right">0.071</td>
</tr>
<tr class="even">
<td align="left">CID4290</td>
<td align="right">0.000</td>
<td align="right">0.408</td>
<td align="right">0.417</td>
<td align="right">0.068</td>
<td align="right">0.072</td>
<td align="right">0.005</td>
<td align="right">0.000</td>
<td align="right">0.024</td>
<td align="right">0.006</td>
</tr>
<tr class="odd">
<td align="left">CID4398</td>
<td align="right">0.000</td>
<td align="right">0.220</td>
<td align="right">0.143</td>
<td align="right">0.119</td>
<td align="right">0.006</td>
<td align="right">0.204</td>
<td align="right">0.000</td>
<td align="right">0.034</td>
<td align="right">0.275</td>
</tr>
<tr class="even">
<td align="left">CID44041</td>
<td align="right">0.332</td>
<td align="right">0.161</td>
<td align="right">0.000</td>
<td align="right">0.122</td>
<td align="right">0.006</td>
<td align="right">0.264</td>
<td align="right">0.000</td>
<td align="right">0.032</td>
<td align="right">0.084</td>
</tr>
<tr class="odd">
<td align="left">CID4461</td>
<td align="right">0.319</td>
<td align="right">0.138</td>
<td align="right">0.292</td>
<td align="right">0.084</td>
<td align="right">0.028</td>
<td align="right">0.000</td>
<td align="right">0.003</td>
<td align="right">0.020</td>
<td align="right">0.116</td>
</tr>
<tr class="even">
<td align="left">CID4463</td>
<td align="right">0.033</td>
<td align="right">0.123</td>
<td align="right">0.560</td>
<td align="right">0.069</td>
<td align="right">0.067</td>
<td align="right">0.036</td>
<td align="right">0.000</td>
<td align="right">0.039</td>
<td align="right">0.074</td>
</tr>
<tr class="odd">
<td align="left">CID4465</td>
<td align="right">0.226</td>
<td align="right">0.287</td>
<td align="right">0.066</td>
<td align="right">0.099</td>
<td align="right">0.070</td>
<td align="right">0.206</td>
<td align="right">0.004</td>
<td align="right">0.013</td>
<td align="right">0.029</td>
</tr>
<tr class="even">
<td align="left">CID4471</td>
<td align="right">0.009</td>
<td align="right">0.288</td>
<td align="right">0.000</td>
<td align="right">0.182</td>
<td align="right">0.046</td>
<td align="right">0.296</td>
<td align="right">0.000</td>
<td align="right">0.143</td>
<td align="right">0.035</td>
</tr>
<tr class="odd">
<td align="left">CID4495</td>
<td align="right">0.366</td>
<td align="right">0.066</td>
<td align="right">0.023</td>
<td align="right">0.071</td>
<td align="right">0.066</td>
<td align="right">0.274</td>
<td align="right">0.002</td>
<td align="right">0.005</td>
<td align="right">0.126</td>
</tr>
<tr class="even">
<td align="left">CID44971</td>
<td align="right">0.286</td>
<td align="right">0.099</td>
<td align="right">0.000</td>
<td align="right">0.083</td>
<td align="right">0.002</td>
<td align="right">0.300</td>
<td align="right">0.000</td>
<td align="right">0.010</td>
<td align="right">0.219</td>
</tr>
<tr class="odd">
<td align="left">CID4513</td>
<td align="right">0.001</td>
<td align="right">0.398</td>
<td align="right">0.000</td>
<td align="right">0.143</td>
<td align="right">0.397</td>
<td align="right">0.000</td>
<td align="right">0.001</td>
<td align="right">0.056</td>
<td align="right">0.005</td>
</tr>
<tr class="even">
<td align="left">CID4515</td>
<td align="right">0.456</td>
<td align="right">0.185</td>
<td align="right">0.000</td>
<td align="right">0.092</td>
<td align="right">0.050</td>
<td align="right">0.131</td>
<td align="right">0.002</td>
<td align="right">0.000</td>
<td align="right">0.085</td>
</tr>
<tr class="odd">
<td align="left">CID4523</td>
<td align="right">0.000</td>
<td align="right">0.219</td>
<td align="right">0.381</td>
<td align="right">0.165</td>
<td align="right">0.025</td>
<td align="right">0.166</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.044</td>
</tr>
<tr class="even">
<td align="left">CID4530</td>
<td align="right">0.085</td>
<td align="right">0.317</td>
<td align="right">0.284</td>
<td align="right">0.083</td>
<td align="right">0.034</td>
<td align="right">0.068</td>
<td align="right">0.000</td>
<td align="right">0.040</td>
<td align="right">0.088</td>
</tr>
<tr class="odd">
<td align="left">CID4535</td>
<td align="right">0.075</td>
<td align="right">0.062</td>
<td align="right">0.681</td>
<td align="right">0.049</td>
<td align="right">0.007</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
<td align="right">0.029</td>
<td align="right">0.097</td>
</tr>
<tr class="even">
<td align="left">CID4040</td>
<td align="right">0.305</td>
<td align="right">0.217</td>
<td align="right">0.272</td>
<td align="right">0.052</td>
<td align="right">0.030</td>
<td align="right">0.000</td>
<td align="right">0.002</td>
<td align="right">0.009</td>
<td align="right">0.114</td>
</tr>
<tr class="odd">
<td align="left">CID3838</td>
<td align="right">0.169</td>
<td align="right">0.200</td>
<td align="right">0.332</td>
<td align="right">0.085</td>
<td align="right">0.058</td>
<td align="right">0.077</td>
<td align="right">0.000</td>
<td align="right">0.027</td>
<td align="right">0.052</td>
</tr>
<tr class="even">
<td align="left">CID3946</td>
<td align="right">0.389</td>
<td align="right">0.147</td>
<td align="right">0.000</td>
<td align="right">0.081</td>
<td align="right">0.038</td>
<td align="right">0.256</td>
<td align="right">0.000</td>
<td align="right">0.004</td>
<td align="right">0.085</td>
</tr>
</tbody>
</table>
<p>We can also visualise the results as a barplot trough the built-in <code>plot_deconvolution</code> function</p>
<pre><code><span><span class="co">#&gt; Error in path.expand(path): invalid 'path' argument</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="deconvolution-with-bayesprism">6.2 Deconvolution with BayesPrism<a class="anchor" aria-label="anchor" href="#deconvolution-with-bayesprism"></a>
</h3>
<p>The third method we will use is <a href="https://github.com/Danko-Lab/BayesPrism" class="external-link">BayesPrism</a> <span class="citation">(Chu et al. <a href="#ref-Chu2022" role="doc-biblioref">2022</a>)</span>. This method is based on a Bayesian framework and models the transcriptomic expression observed in the scRNA-seq dataset. It then uses this information to dissect te bulk RNA-seq.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BayesPrism deconvolution</span></span>
<span></span>
<span><span class="va">deconvolution.results.bayesprism</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span>bulk_gene_expression <span class="op">=</span> <span class="va">bulk.data</span>,</span>
<span>                                           single_cell_object <span class="op">=</span> <span class="va">counts.matrix</span>,</span>
<span>                                           cell_type_annotations <span class="op">=</span> <span class="va">cell.type.annotations</span>,</span>
<span>                                           signature<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                           method <span class="op">=</span> <span class="st">'bayesprism'</span>, </span>
<span>                                           n_cores<span class="op">=</span><span class="fl">12</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the results as before:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_deconvolution.html">plot_deconvolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'bayesprism'</span> <span class="op">=</span> <span class="va">deconvolution.results.bayesprism</span><span class="op">)</span>, <span class="st">"bar"</span>, <span class="st">"method"</span>, <span class="st">"Spectral"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in path.expand(path): invalid 'path' argument</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="deconvolution-of-the-bulk-data-at-a-lower-resolution">7. Deconvolution of the bulk data at a lower resolution<a class="anchor" aria-label="anchor" href="#deconvolution-of-the-bulk-data-at-a-lower-resolution"></a>
</h2>
<p>The considered single-cell breast cancer dataset includes cell-type annotations at three levels of resolution: <code>celltype_major</code>, <code>celltype_minor</code>, and <code>celltype_subset</code>, which distinguish 9, 29, and 58 cell types respectively. The different cell-type annotations can be accessed with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single.cell.data</span><span class="op">$</span><span class="va">celltype_major</span>      <span class="co"># Major annotation</span></span>
<span><span class="va">single.cell.data</span><span class="op">$</span><span class="va">celltype_minor</span>      <span class="co"># Minor annotation</span></span>
<span><span class="va">single.cell.data</span><span class="op">$</span><span class="va">celltype_subset</span>     <span class="co"># Subset annotation</span></span></code></pre></div>
<p>These additional annotations provide a cell-type classification at a finer resolution. For instance, at the <code>celltype_major</code> level, we only have the T cell population, while at the <code>celltype_minor</code> level, we can distinguish between CD4+ and CD8+ T cells. In the following, we will again perform deconvolution analysis with DWLS but, this time, using the <code>celltype_minor</code> information. We will subsample the dataset as before, this time considering the second level of resolution for the cell types, and extract the objects needed for deconvolution.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_cells_per_celltype</span> <span class="op">=</span> <span class="fl">200</span></span>
<span></span>
<span></span>
<span><span class="va">sampled.metadata</span> <span class="op">&lt;-</span> <span class="va">single.cell.data</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">rownames_to_column</span><span class="op">(</span><span class="va">.</span>, <span class="st">'barcode'</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">group_by</span><span class="op">(</span><span class="va">.</span>, <span class="va">celltype_minor</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">nest</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>            </span>
<span>      <span class="fu">mutate</span><span class="op">(</span>n <span class="op">=</span>  <span class="fu">map_dbl</span><span class="op">(</span><span class="va">data</span>, <span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">mutate</span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">n</span>, <span class="va">max_cells_per_celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">mutate</span><span class="op">(</span>samp <span class="op">=</span> <span class="fu">map2</span><span class="op">(</span><span class="va">data</span>, <span class="va">n</span>, <span class="va">sample_n</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>      <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu">unnest</span><span class="op">(</span><span class="va">samp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single.cell.data.sampled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">single.cell.data</span>, cells <span class="op">=</span> <span class="va">sampled.metadata</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Number of cells per cell type (minor resolution) after subsampling</caption>
<thead><tr class="header">
<th align="left">celltype_minor</th>
<th align="right">number_cells</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B cells Memory</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">B cells Naive</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">CAFs MSC iCAF-like</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">CAFs myCAF-like</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Cancer Basal SC</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Cancer Cycling</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Cancer Her2 SC</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Cancer LumA SC</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Cancer LumB SC</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Cycling PVL</td>
<td align="right">37</td>
</tr>
<tr class="odd">
<td align="left">Cycling T-cells</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Cycling_Myeloid</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">DCs</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Endothelial ACKR1</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Endothelial CXCL12</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Endothelial Lymphatic LYVE1</td>
<td align="right">183</td>
</tr>
<tr class="odd">
<td align="left">Endothelial RGS5</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Luminal Progenitors</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Macrophage</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Mature Luminal</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Monocyte</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">Myoepithelial</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">NK cells</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">NKT cells</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">Plasmablasts</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">PVL Differentiated</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">PVL Immature</td>
<td align="right">200</td>
</tr>
<tr class="even">
<td align="left">T cells CD4+</td>
<td align="right">200</td>
</tr>
<tr class="odd">
<td align="left">T cells CD8+</td>
<td align="right">200</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">single.cell.data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">cell.type.annotations</span> <span class="op">&lt;-</span> <span class="va">single.cell.data</span><span class="op">$</span><span class="va">celltype_minor</span></span>
<span><span class="va">batch.ids</span> <span class="op">&lt;-</span> <span class="va">single.cell.data</span><span class="op">$</span><span class="va">orig.ident</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">signature.matrix.dwls.minor</span> <span class="op">&lt;-</span> <span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="../reference/build_model.html">build_model</a></span><span class="op">(</span>single_cell_object <span class="op">=</span> <span class="va">counts.matrix</span>,</span>
<span>                        cell_type_annotations <span class="op">=</span> <span class="va">cell.type.annotations</span>,</span>
<span>                        method <span class="op">=</span> <span class="st">'dwls'</span>, </span>
<span>                        dwls_method <span class="op">=</span> <span class="st">'mast_optimized'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">deconvolution.results.dwls.minor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvolute.html">deconvolute</a></span><span class="op">(</span>bulk_gene_expression <span class="op">=</span> <span class="va">bulk.data</span>,</span>
<span>                                          signature <span class="op">=</span> <span class="va">signature.matrix.dwls.minor</span>,</span>
<span>                                          method<span class="op">=</span><span class="st">'dwls'</span>,</span>
<span>                                          dwls_submethod <span class="op">=</span> <span class="st">'DampenedWLS'</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the results as before:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">omnideconv</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_deconvolution.html">plot_deconvolution</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'dwls'</span> <span class="op">=</span> <span class="va">deconvolution.results.dwls.minor</span><span class="op">)</span>, <span class="st">"bar"</span>, <span class="st">"method"</span>, <span class="st">"Spectral"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in path.expand(path): invalid 'path' argument</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comparison-of-cell-fractions-across-conditions">8. Comparison of cell fractions across conditions<a class="anchor" aria-label="anchor" href="#comparison-of-cell-fractions-across-conditions"></a>
</h2>
<p>We can consider as well the metadata provided with the original paper, which include patient’s data, cancer subtype information and treatment details. We can first harmonize the sample names, and then combine all this information with the deconvolution results in one dataframe.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">patient.metadata</span> <span class="op">&lt;-</span> <span class="fu">read_excel</span><span class="op">(</span><span class="st">"C:/Users/c1041161/book_chapter/41588_2021_911_MOESM4_ESM.xlsx"</span>, sheet <span class="op">=</span> <span class="fl">1</span>, skip <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">patient.metadata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Sample'</span>, <span class="st">'Grade'</span>, <span class="st">'Cancer_type'</span>, <span class="st">'IHC_subtype'</span>, <span class="st">'Treatment'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">patient.metadata</span><span class="op">$</span><span class="va">Sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'-'</span>, <span class="st">''</span>, <span class="va">patient.metadata</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'CID'</span>, <span class="va">.</span><span class="op">)</span></span>
<span></span>
<span><span class="va">patients.results</span> <span class="op">&lt;-</span> <span class="fu">rownames_to_column</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">deconvolution.results.dwls</span><span class="op">)</span>, <span class="st">'Sample'</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">gather</span><span class="op">(</span><span class="va">.</span>, key<span class="op">=</span><span class="st">'celltype'</span>, value<span class="op">=</span><span class="st">'cell_fraction'</span>, <span class="op">-</span><span class="va">Sample</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">.</span>, <span class="va">patient.metadata</span><span class="op">)</span></span></code></pre></div>
<p>Each condition has a different number of samples:</p>
<p>We can visualize the estimated cell fractions across samples with a boxplot, and group the results either by IHC subtype or by treatment status</p>
<div class="figure">
<img src="figure/unnamed-chunk-27-1.png" alt=""><p class="caption">plot of chunk unnamed-chunk-27</p>
</div>
<p>HER2-positive samples are enriched in T cells, while the ER-positive and triple negative breast cancer (TNBC) samples show an enrichment in cancer and normal epithelial cells, respectively. The patient metadata includes information about eventual treatments undergone by the patients. We can see that 5 out of the 24 patients were treated with Neoadjuvant and/or Paclitaxel, while the other 19 were untreated. We can therefore visualize the distribution of cell fractions across treated and untreated patients</p>
<div class="figure">
<img src="figure/unnamed-chunk-28-1.png" alt=""><p class="caption">plot of chunk unnamed-chunk-28</p>
</div>
<p>The major differences in this case are the enrichment of the CAFs and normal epithelial cells in the treated samples, as opposed to the cancer epithelial cells and B cells which have a higher median cell fraction in the untreated samples.</p>
<p>As mentioned before, with the lower level of annotations we can identify additional cell types, such as CD4+/CD8+ subtypes, Fibroblasts and cancer cells. We can therefore visualize the estimated cell fractions in a boxplot, focusing in particular on the CD4+ and CD8+ T-cell subtypes.</p>
<div class="figure">
<img src="figure/unnamed-chunk-29-1.png" alt=""><p class="caption">plot of chunk unnamed-chunk-29</p>
</div>
<p>The estimates for the CD8+ T cells seem to be very low for all samples. On the other hand, in the HER2/ER positive samples the cell fractions estimated for the CD4+ T cells seem to be significantly higher than in the other subtypes. The estimates for the CD8+ T cells are close to zero for almost every sample, with very low fractions detected for the HER2+/ER+ ones.</p>
<p>Similarly, we can visualize the distribution of the estimates for the cancer associated fibroblasts (CAFs), respectively the myoblastic CAFs (myCAF) and the bone marrow derived inflammatory CAFs (MSC iCAF).</p>
<div class="figure">
<img src="figure/unnamed-chunk-30-1.png" alt=""><p class="caption">plot of chunk unnamed-chunk-30</p>
</div>
<p>Here we can notice that, while the myoblastic CAFs have a comparable median value, the MSC CAFs have higher infiltration in the samples characterized as HER2+/ER+ subtype.</p>
<p>Finally, we can visualize the distribution of the different cancer cells molecular subtypes that were described by the authors.</p>
<div class="figure">
<img src="figure/unnamed-chunk-31-1.png" alt=""><p class="caption">plot of chunk unnamed-chunk-31</p>
</div>
<p>We can see that the HER2+ shows an overrepresentation of the cancer cells described as HER2-like by the authors. Similarly, the ER+ samples show higher fractions of Luminal B cancer cells. These findings on bulk RNA-seq are concordant with the subtypings that the authors described on the single cell RNA-seq (‘Results - scSubtype’).</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-Chu2022">
<p>Chu, Tinyi, Zhong Wang, Dana Pe’er, and Charles G. Danko. 2022. “Cell Type and Gene Expression Deconvolution with Bayesprism Enables Bayesian Integrative Analysis Across Bulk and Single-Cell Rna Sequencing in Oncology.” <em>Nature Cancer</em> 3 (4): 505–17. <a href="https://doi.org/10.1038/s43018-022-00356-3" class="external-link">https://doi.org/10.1038/s43018-022-00356-3</a>.</p>
</div>
<div id="ref-Finak2015">
<p>Finak, Greg, Andrew McDavid, Masanao Yajima, Jingyuan Deng, Vivian Gersuk, Alex K. Shalek, Chloe K. Slichter, et al. 2015. “MAST: A Flexible Statistical Framework for Assessing Transcriptional Changes and Characterizing Heterogeneity in Single-Cell Rna Sequencing Data.” <em>Genome Biology</em> 16 (1). <a href="https://doi.org/10.1186/s13059-015-0844-5" class="external-link">https://doi.org/10.1186/s13059-015-0844-5</a>.</p>
</div>
<div id="ref-Hao2023">
<p>Hao, Yuhan, Tim Stuart, Madeline H. Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, et al. 2023. “Dictionary Learning for Integrative, Multimodal and Scalable Single-Cell Analysis.” <em>Nature Biotechnology</em> 42 (2): 293–304. <a href="https://doi.org/10.1038/s41587-023-01767-y" class="external-link">https://doi.org/10.1038/s41587-023-01767-y</a>.</p>
</div>
<div id="ref-Heumos2023">
<p>Heumos, Lukas, Anna C. Schaar, Christopher Lance, Anastasia Litinetskaya, Felix Drost, Luke Zappia, Malte D. Lücken, et al. 2023. “Best Practices for Single-Cell Analysis Across Modalities.” <em>Nature Reviews Genetics</em> 24 (8): 550–72. <a href="https://doi.org/10.1038/s41576-023-00586-w" class="external-link">https://doi.org/10.1038/s41576-023-00586-w</a>.</p>
</div>
<div id="ref-McDavid2012">
<p>McDavid, Andrew, Greg Finak, Pratip K. Chattopadyay, Maria Dominguez, Laurie Lamoreaux, Steven S. Ma, Mario Roederer, and Raphael Gottardo. 2012. “Data Exploration, Quality Control and Testing in Single-Cell qPCR-Based Gene Expression Experiments.” <em>Bioinformatics</em> 29 (4): 461–67. <a href="https://doi.org/10.1093/bioinformatics/bts714" class="external-link">https://doi.org/10.1093/bioinformatics/bts714</a>.</p>
</div>
<div id="ref-Nault2022">
<p>Nault, Rance, Satabdi Saha, Sudin Bhattacharya, Jack Dodson, Samiran Sinha, Tapabrata Maiti, and Tim Zacharewski. 2022. “Benchmarking of a Bayesian Single Cell Rnaseq Differential Gene Expression Test for Dose–Response Study Designs.” <em>Nucleic Acids Research</em> 50 (8): e48–e48. <a href="https://doi.org/10.1093/nar/gkac019" class="external-link">https://doi.org/10.1093/nar/gkac019</a>.</p>
</div>
<div id="ref-Tsoucas2019">
<p>Tsoucas, Daphne, Rui Dong, Haide Chen, Qian Zhu, Guoji Guo, and Guo-Cheng Yuan. 2019. “Accurate Estimation of Cell-Type Composition from Gene Expression Data.” <em>Nature Communications</em> 10 (1). <a href="https://doi.org/10.1038/s41467-019-10802-z" class="external-link">https://doi.org/10.1038/s41467-019-10802-z</a>.</p>
</div>
<div id="ref-Wu2021">
<p>Wu, Sunny Z., Ghamdan Al-Eryani, Daniel Lee Roden, Simon Junankar, Kate Harvey, Alma Andersson, Aatish Thennavan, et al. 2021. “A Single-Cell and Spatially Resolved Atlas of Human Breast Cancers.” <em>Nature Genetics</em> 53 (9): 1334–47. <a href="https://doi.org/10.1038/s41588-021-00911-1" class="external-link">https://doi.org/10.1038/s41588-021-00911-1</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lorenzo Merotto, Alexander Dietrich, Konstantin Pelz, Nicolas Goedert, Yuyu Liang, Katharina Reinisch.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
