#' Setting the CibersortX Credentials
#'
#' @param email The email which was used to register on the website.
#' @param token The token provided by the website.
#'
#' @export
#'
#' @examples
set_cibersortx_credentials = function(email,token) {
  assign("cibersortx_email", email, envir=config_env)
  assign("cibersortx_token", token, envir=config_env)
}


#' Signature matrix creation with CibersortX
#'
#' @param single_cell_object A matrix or dataframe with the single-cell data. Rows are genes, columns are samples. Row and column names need to be set.
#' @param cell_type_annotations A Vector of the cell type annotations. Has to be in the same order as the samples in single_cell_object
#' @param verbose Whether CibersortX should print out what it is doing.
#' @param input_dir The directory in which the input files can be found (or are created in). Default is a temporary directory.
#' @param output_dir The directory in which the output files are saved. Default is a temporary directory.
#' @param display_heatmap Whether to display the generated heatmap.
#' @param k_max Maximum condition number (default: 999). Will be added to the file name.
#' @param ... Additional parameters supplied to the algorithm. Options are:
#' G_min: <int> Minimum number of genes per cell type in sig. matrix (default: 300)
#' G_max: <int> Maximum number of genes per cell type in sig. matrix (default: 500)
#' q_value: <int> Q-value threshold for differential expression (default: 0.01)
#' filter: <bool> Remove non-hematopoietic genes (default: FALSE)
#' remake: <bool> Remake signature gene matrix (default: False)
#' replicates: <int> Number of replicates to use for building scRNAseq reference file (default: 5)
#' sampling: <float> Fraction of available single cell GEPs selected using random sampling (default: 0.5)
#' fraction: <float> Fraction of cells of same identity showing evidence of expression (default: 0.75)
#'
#'
#' @return The signature matrix. Rows are genes, columns are cell types.
#' @export
#'
#' @examples
cibersort_generate_signature <- function(single_cell_object, cell_type_annotations, verbose = TRUE, input_dir = NULL, output_dir = NULL, display_heatmap = FALSE, k_max=999, ...){
  if (docker_available()){
    temp_dir <- tempdir()
    if (is.null(input_dir)){
      input_dir <- temp_dir
    }
    if (is.null(output_dir)){
      output_dir <- temp_dir
    }

    if (class(single_cell_object)[1]!="character"){
      transform_and_save_single_cell(single_cell_object,cell_type_annotations,input_dir,verbose)
      single_cell_object_filename <- "sample_file_for_cibersort.txt"
    } else {
      single_cell_object_filename <- single_cell_object
    }
    command_to_run <- create_docker_command(input_dir, output_dir, method = "create_sig", verbose = verbose, refsample=single_cell_object_filename, k_max=k_max)

    if (!verbose){
      if (Sys.info()['sysname']=="Windows"){
        base::message("The windows implementation requires verbose mode. It is now switched on.")
        verbose <- TRUE
      }
    }

    filebase <- paste0("CIBERSORTx_sample_file_for_cibersort_inferred_phenoclasses.CIBERSORTx_sample_file_for_cibersort_inferred_refsample.bm.K",k_max)
    filename_sig_matrix <- paste0(filebase,".txt")
    full_path <- paste0(input_dir,"/",filename_sig_matrix)

    if (file.exists(full_path))
      file.remove(full_path)

    code <- system(command_to_run, ignore.stdout = !verbose, ignore.stderr = !verbose)
    if (code!=0){
      base::message(paste("Something went wrong: Error code ",code,". Please try again with \"verbose=TRUE\""))
    }

    if (display_heatmap){
      filename_heatmap <- paste0(filebase,".pdf")
      Biobase::openPDF(normalizePath(paste0(output_dir,"/",filename_heatmap)))
    }

    sig_matrix <- readr::read_tsv(paste0(output_dir,"/",filename_sig_matrix))
    rownames(sig_matrix)<-sig_matrix$NAME

    return(as.matrix.data.frame(sig_matrix[,-1]))

  } else {
    print("Installation of docker can not be found. Please check whether you can
          call 'docker' in the command line and get a help menu")
  }
}

#' Deconvolute with CibersortX
#'
#' @param bulk_gene_expression A matrix or dataframe with the bulk data. Rows are genes, columns are samples.
#' Can also just be the filename of the bulk expression data in the correct format (generated by using this
#' method). The file needs to be in the folder "input_dir".
#' @param signature The signature matrix. Can also be the filename of the signature matrix in the correct
#' format (generated by using this method). The file needs to be in the folder "input_dir".
#' @param verbose Whether to output what CibersortX is doing.
#' @param input_dir The folder in which the input files can be found (or are created in). Default is a temporary directory.
#' @param output_dir The directory in which the output files are saved. Default is a temporary directory.
#' @param display_extra_info Whether to print the "P.value","Correlation" and "RMSE" calculated by CibersortX.
#' @param label The label which will be added to the file name. Default is "none", adding no label at all.
#' @param ... Additional parameters supplied to the algorithm. Options are:
#' perm: <int> No. of permutations for p-value calculation (default: 0)
#' rmbatch_B_mode: <bool>  Run B-mode batch correction (default: FALSE)
#' rmbatch_S_mode: <bool>  Run S-mode batch correction (default: FALSE)
#' source_GEPs: <file_name>  Signature matrix GEPs for batch correction (default: sigmatrix)
#' qn: <bool>  Run quantile normalization (default: FALSE)
#' absolute: <bool>  Run absolute mode (default: FALSE)
#' abs_method: <char>  Pick absolute method ("sig.score" (default) or "no.sumto1")
#'
#'
#' @return A matrix with the probabilities of each cell-type for each individual. Rows are individuals, columns are cell types.
#' @export
#'
#' @examples
deconvolute_cibersort <- function(bulk_gene_expression, signature, verbose = TRUE, input_dir = NULL, output_dir = NULL, display_extra_info=FALSE , label="none", ...){
  if (docker_available()){
    temp_dir <- tempdir()
    if (is.null(input_dir))
      input_dir <- temp_dir
    if (is.null(output_dir))
      output_dir <- temp_dir

    if (class(signature)[1]!="character"){
      sig <- paste0(input_dir,"/signature_matrix.txt")
      readr::write_tsv(data.frame("NAME"=rownames(signature),signature),sig)
      sigmatrix_filename <- "signature_matrix.txt"
    } else {
      sigmatrix_filename <- signature
    }
    if (class(bulk_gene_expression)[1]!="character"){
      transform_and_save_bulk(bulk_gene_expression,input_dir,verbose)
      bulk_gene_expression_filename <- "mixture_file_for_cibersort.txt"
    } else {
      bulk_gene_expression_filename <- bulk_gene_expression
    }
    filename_cell_props <- paste0("CIBERSORTx_",label,"_Results.txt")
    command_to_run <- create_docker_command(input_dir,output_dir,method = "impute_cell_fractions",verbose = verbose,
                                            sigmatrix=sigmatrix_filename, mixture <- bulk_gene_expression_filename, label = label)

    if (!verbose){
      if (Sys.info()['sysname']=="Windows"){
        base::message("The windows implementation requires verbose mode. It is now switched on.")
        verbose <- TRUE
      }
    }

    code <- system(command_to_run, ignore.stdout = !verbose, ignore.stderr = !verbose)
    if (code!=0){
      base::message(paste("Something went wrong: Error code ",code,". Please try again with \"verbose=TRUE\""))
    }

    cell_props <- readr::read_tsv(paste0(output_dir,"/",filename_cell_props))
    rownames(cell_props)<-cell_props$Mixture
    cell_props <- cell_props[,-1]

    extra_cols <- c("P.value","Correlation","RMSE")
    if (display_extra_info){
      print(cell_props[,extra_cols])
    }

    return(as.matrix.data.frame(cell_props[,! names(cell_props) %in% extra_cols]))

  } else {
    print("Installation of docker can not be found. Please check whether you can
          call 'docker' in the command line and get a help menu")
  }
}

transform_and_save_single_cell <- function(sc_matrix,cell_types,path,verbose=TRUE){
  colnames(sc_matrix)<-cell_types
  output <- rbind(colnames(sc_matrix),sc_matrix)
  rownames(output)<-c("GeneSymbol",rownames(sc_matrix))
  output <- data.frame("GeneSymbol"= rownames(output),output)
  output_file <- paste0(path,"/sample_file_for_cibersort.txt")
  readr::write_tsv(output,output_file, col_names = FALSE)
  if (verbose){
    base::message(paste("Single cell matrix was saved successfully and can be found at: ",output_file))
  }
  return(output_file)
}

transform_and_save_bulk <- function(bulk, path,verbose=TRUE){
  output_file <- paste0(path,"/mixture_file_for_cibersort.txt")
  readr::write_tsv(data.frame("Gene"=rownames(bulk),bulk),output_file)
  if (verbose){
    base::message(paste("Bulk data matrix was saved successfully and can be found at: ",output_file))
  }
  return(output_file)
}

create_docker_command <- function(in_dir, out_dir, method = c("create_sig","impute_cell_fractions"),verbose = TRUE, ...){
  base <- paste0("docker run -v ",in_dir,":/src/data:z -v ",out_dir,":/src/outdir:z cibersortx/fractions --single_cell TRUE")
  if (verbose){
    base <- paste(base,"--verbose TRUE")
  }
  assertthat::assert_that(exists("cibersortx_email", envir=config_env), msg="CibersortX email for credentials is missing. Please call set_cibersortx_credentials(email,token) first.")
  assertthat::assert_that(exists("cibersortx_token", envir=config_env), msg="CibersortX token for credentials is missing. Please call set_cibersortx_credentials(email,token) first.")
  credentials <- paste("--username",get("cibersortx_email", envir=config_env),"--token",get("cibersortx_token", envir=config_env))
  return(paste(base,credentials,get_method_options(method,...)))
}

get_method_options <- function(method,...){
  if (method=="create_sig"){
    return(get_signature_matrix_options(...))
  } else if (method == "impute_cell_fractions"){
    return(get_cell_fractions_options(...))
  } else {
    base::stop(paste("Method",method,"is not valid"))
  }
}

get_signature_matrix_options <- function(refsample, G_min = 300, G_max = 500, q_value = 0.01, filter = FALSE,
                                         k_max = 999, remake = FALSE, replicates = 5, sampling = 0.5, fraction = 0.75){
  return(paste("--refsample",refsample,"--G.min",G_min,"--G.max",G_max,"--q.value",q_value,"--filter",filter,
               "--k.max",k_max,"--remake",remake,"--replicates",replicates,"--sampling",sampling,"--fraction",fraction))
}


get_cell_fractions_options <- function(sigmatrix,mixture,perm=0,label="none",rmbatch_B_mode=FALSE,rmbatch_S_mode=FALSE,
                                       source_GEPs=sigmatrix,qn=FALSE,absolute=FALSE,abs_method="sig.score"){
  return(paste("--mixture",mixture,"--sigmatrix",sigmatrix,"--perm",perm,"--label",label,"--rmbatchBmode",rmbatch_B_mode,"--rmbatchSmode",rmbatch_S_mode,
               "--sourceGEPs",source_GEPs,"QN",qn,"--absolute",absolute,"--abs_method",abs_method))
}

#Working Commands
#docker run -v /c/Users/Konstantin/Desktop/Uni/7Semester/SysBioMed/testIn:/src/data -v /c/Users/Konstantin/Desktop/Uni/7Semester/SysBioMed/testOut:/src/outdir cibersortx/fractions --username mailto:gregor.sturm@cs.tum.edu --token 1860fb7bc414a958e8aa1e91a5229d8a --single_cell TRUE --refsample sampleFileForCibersort.txt
#docker run -v /c/Users/Konstantin/Desktop/Uni/7Semester/SysBioMed/testIn:/src/data -v /c/Users/Konstantin/Desktop/Uni/7Semester/SysBioMed/testOut:/src/outdir cibersortx/fractions --username konstantin.pelz@tum.de --token 27308ae0ef1458d381becac46ca7e480 --single_cell TRUE --refsample sampleFileForCibersort.txt




